{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import bgen_reader\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "filedir = '/vol/bmd/meliao/data/haplotype/hap/'\n",
    "chr_num = '22'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/vol/bmd/yanyul/miniconda3/envs/haplotype_po/lib/python3.7/site-packages/bgen_reader/_reader.py:135: UserWarning: You don't have permission to write to `/vol/bmd/meliao/data/haplotype/hap/ukb_hap_chr22_v2.bgen.metadata`.\n",
      "This might prevent speeding-up the reading process in future runs.\n",
      "\n",
      "  warnings.warn(_metafile_nowrite_dir.format(filepath=fp), UserWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We will create the metafile `/tmp/tmps173syr2/%vol%bmd%meliao%data%haplotype%hapukb_hap_chr22_v2.bgen.metadata`. This file will speed up further\n",
      "reads and only need to be created once. So, please, bear with me.\n",
      "Sample IDs are read from /vol/bmd/meliao/data/haplotype/link_files/ukb1952_v2_s487398.sample.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping variants: 100%|██████████| 10911/10911 [00:00<00:00, 67679.48it/s]\n"
     ]
    }
   ],
   "source": [
    "bgen = bgen_reader.read_bgen(\n",
    "    f'{filedir}/ukb_hap_chr{chr_num}_v2.bgen', \n",
    "    samples_filepath = '/vol/bmd/meliao/data/haplotype/link_files/ukb1952_v2_s487398.sample'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas = pd.read_csv(\n",
    "    '/vol/bmd/yanyul/UKB/haplotype_imputation/20002_1065.gwas.imputed_v3.both_sexes.tsv.bgz', \n",
    "    header = 0, \n",
    "    sep = '\\t', \n",
    "    compression = 'gzip'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_map = pd.read_csv(\n",
    "    '/vol/bmd/yanyul/UKB/haplotype_imputation/snp_map_for_neale_lab_gwas.tsv.gz',\n",
    "    sep='\\t',\n",
    "    header=0,\n",
    "    compression='gzip',\n",
    "    dtype={3:'str'}\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Workflow\n",
    "\n",
    "1. **In GWAS**, extract variants that occur in my genotype.\n",
    "2. **In GWAS**, perform LD clumping.\n",
    "3. Calculate PRS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_map = snp_map[ (snp_map['assigned_id'] != 'not_shown') & (snp_map['assigned_id'] != 'ambiguous') ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract variants from GWAS\n",
    "gwas_sub = gwas[ gwas['variant'].isin(snp_map['assigned_id']) ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>expected_case_minor_AC</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1:754105:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.036358</td>\n",
       "      <td>6803.38</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>26261.0</td>\n",
       "      <td>6774.0</td>\n",
       "      <td>-0.000939</td>\n",
       "      <td>0.002727</td>\n",
       "      <td>-0.344463</td>\n",
       "      <td>0.730499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1:756604:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.129866</td>\n",
       "      <td>24300.60</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>628482.0</td>\n",
       "      <td>162876.0</td>\n",
       "      <td>0.000781</td>\n",
       "      <td>0.001491</td>\n",
       "      <td>0.524009</td>\n",
       "      <td>0.600273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>1:768448:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.105920</td>\n",
       "      <td>19819.70</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>76504.0</td>\n",
       "      <td>19920.0</td>\n",
       "      <td>0.001463</td>\n",
       "      <td>0.001626</td>\n",
       "      <td>0.899687</td>\n",
       "      <td>0.368287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>1:779322:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.127848</td>\n",
       "      <td>23922.80</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>92342.0</td>\n",
       "      <td>23900.0</td>\n",
       "      <td>-0.000413</td>\n",
       "      <td>0.001500</td>\n",
       "      <td>-0.275184</td>\n",
       "      <td>0.783175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>1:801536:T:G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.015058</td>\n",
       "      <td>2817.62</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>10876.0</td>\n",
       "      <td>2813.0</td>\n",
       "      <td>-0.000808</td>\n",
       "      <td>0.004117</td>\n",
       "      <td>-0.196378</td>\n",
       "      <td>0.844315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364106</th>\n",
       "      <td>22:51177257:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.037381</td>\n",
       "      <td>6994.83</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>27000.0</td>\n",
       "      <td>7086.0</td>\n",
       "      <td>0.002953</td>\n",
       "      <td>0.002644</td>\n",
       "      <td>1.117070</td>\n",
       "      <td>0.263965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364151</th>\n",
       "      <td>22:51183255:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.334182</td>\n",
       "      <td>62532.20</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>480908.0</td>\n",
       "      <td>124984.0</td>\n",
       "      <td>0.002137</td>\n",
       "      <td>0.001062</td>\n",
       "      <td>2.013180</td>\n",
       "      <td>0.044097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364161</th>\n",
       "      <td>22:51185848:G:A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.050516</td>\n",
       "      <td>9452.61</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>685795.0</td>\n",
       "      <td>177679.0</td>\n",
       "      <td>0.001040</td>\n",
       "      <td>0.002284</td>\n",
       "      <td>0.455429</td>\n",
       "      <td>0.648801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364198</th>\n",
       "      <td>22:51193629:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.076289</td>\n",
       "      <td>14275.20</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>55102.0</td>\n",
       "      <td>14323.0</td>\n",
       "      <td>0.001454</td>\n",
       "      <td>0.001885</td>\n",
       "      <td>0.771293</td>\n",
       "      <td>0.440534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13364273</th>\n",
       "      <td>22:51217954:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.073265</td>\n",
       "      <td>13709.30</td>\n",
       "      <td>False</td>\n",
       "      <td>361141</td>\n",
       "      <td>52918.0</td>\n",
       "      <td>13737.0</td>\n",
       "      <td>0.001297</td>\n",
       "      <td>0.001922</td>\n",
       "      <td>0.674650</td>\n",
       "      <td>0.499899</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>639250 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  variant minor_allele  minor_AF  expected_case_minor_AC  \\\n",
       "23           1:754105:C:T            T  0.036358                 6803.38   \n",
       "32           1:756604:A:G            A  0.129866                24300.60   \n",
       "55           1:768448:G:A            A  0.105920                19819.70   \n",
       "77           1:779322:A:G            G  0.127848                23922.80   \n",
       "124          1:801536:T:G            G  0.015058                 2817.62   \n",
       "...                   ...          ...       ...                     ...   \n",
       "13364106  22:51177257:C:T            T  0.037381                 6994.83   \n",
       "13364151  22:51183255:A:G            A  0.334182                62532.20   \n",
       "13364161  22:51185848:G:A            G  0.050516                 9452.61   \n",
       "13364198  22:51193629:A:G            G  0.076289                14275.20   \n",
       "13364273  22:51217954:G:A            A  0.073265                13709.30   \n",
       "\n",
       "          low_confidence_variant  n_complete_samples        AC       ytx  \\\n",
       "23                         False              361141   26261.0    6774.0   \n",
       "32                         False              361141  628482.0  162876.0   \n",
       "55                         False              361141   76504.0   19920.0   \n",
       "77                         False              361141   92342.0   23900.0   \n",
       "124                        False              361141   10876.0    2813.0   \n",
       "...                          ...                 ...       ...       ...   \n",
       "13364106                   False              361141   27000.0    7086.0   \n",
       "13364151                   False              361141  480908.0  124984.0   \n",
       "13364161                   False              361141  685795.0  177679.0   \n",
       "13364198                   False              361141   55102.0   14323.0   \n",
       "13364273                   False              361141   52918.0   13737.0   \n",
       "\n",
       "              beta        se     tstat      pval  \n",
       "23       -0.000939  0.002727 -0.344463  0.730499  \n",
       "32        0.000781  0.001491  0.524009  0.600273  \n",
       "55        0.001463  0.001626  0.899687  0.368287  \n",
       "77       -0.000413  0.001500 -0.275184  0.783175  \n",
       "124      -0.000808  0.004117 -0.196378  0.844315  \n",
       "...            ...       ...       ...       ...  \n",
       "13364106  0.002953  0.002644  1.117070  0.263965  \n",
       "13364151  0.002137  0.001062  2.013180  0.044097  \n",
       "13364161  0.001040  0.002284  0.455429  0.648801  \n",
       "13364198  0.001454  0.001885  0.771293  0.440534  \n",
       "13364273  0.001297  0.001922  0.674650  0.499899  \n",
       "\n",
       "[639250 rows x 12 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# gwas_sub.to_csv()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>allele_ids</th>\n",
       "      <th>id</th>\n",
       "      <th>rsid</th>\n",
       "      <th>assigned_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>754105</td>\n",
       "      <td>C,T</td>\n",
       "      <td>1</td>\n",
       "      <td>rs12184325</td>\n",
       "      <td>1:754105:C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>756604</td>\n",
       "      <td>A,G</td>\n",
       "      <td>1</td>\n",
       "      <td>rs3131962</td>\n",
       "      <td>1:756604:A:G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>768448</td>\n",
       "      <td>G,A</td>\n",
       "      <td>1</td>\n",
       "      <td>rs12562034</td>\n",
       "      <td>1:768448:G:A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>779322</td>\n",
       "      <td>A,G</td>\n",
       "      <td>1</td>\n",
       "      <td>rs4040617</td>\n",
       "      <td>1:779322:A:G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>801536</td>\n",
       "      <td>T,G</td>\n",
       "      <td>1</td>\n",
       "      <td>rs79373928</td>\n",
       "      <td>1:801536:T:G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>658714</th>\n",
       "      <td>22</td>\n",
       "      <td>51177257</td>\n",
       "      <td>T,C</td>\n",
       "      <td>rs73174437</td>\n",
       "      <td>rs73174437</td>\n",
       "      <td>22:51177257:C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>658715</th>\n",
       "      <td>22</td>\n",
       "      <td>51183255</td>\n",
       "      <td>A,G</td>\n",
       "      <td>rs5771002</td>\n",
       "      <td>rs5771002</td>\n",
       "      <td>22:51183255:A:G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>658716</th>\n",
       "      <td>22</td>\n",
       "      <td>51185848</td>\n",
       "      <td>G,A</td>\n",
       "      <td>rs3865764</td>\n",
       "      <td>rs3865764</td>\n",
       "      <td>22:51185848:G:A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>658717</th>\n",
       "      <td>22</td>\n",
       "      <td>51193629</td>\n",
       "      <td>G,A</td>\n",
       "      <td>rs142680588</td>\n",
       "      <td>rs142680588</td>\n",
       "      <td>22:51193629:A:G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>658718</th>\n",
       "      <td>22</td>\n",
       "      <td>51217954</td>\n",
       "      <td>A,G</td>\n",
       "      <td>rs9616974</td>\n",
       "      <td>rs9616974</td>\n",
       "      <td>22:51217954:G:A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>639250 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom       pos allele_ids           id         rsid      assigned_id\n",
       "1           1    754105        C,T            1   rs12184325     1:754105:C:T\n",
       "2           1    756604        A,G            1    rs3131962     1:756604:A:G\n",
       "5           1    768448        G,A            1   rs12562034     1:768448:G:A\n",
       "6           1    779322        A,G            1    rs4040617     1:779322:A:G\n",
       "7           1    801536        T,G            1   rs79373928     1:801536:T:G\n",
       "...       ...       ...        ...          ...          ...              ...\n",
       "658714     22  51177257        T,C   rs73174437   rs73174437  22:51177257:C:T\n",
       "658715     22  51183255        A,G    rs5771002    rs5771002  22:51183255:A:G\n",
       "658716     22  51185848        G,A    rs3865764    rs3865764  22:51185848:G:A\n",
       "658717     22  51193629        G,A  rs142680588  rs142680588  22:51193629:A:G\n",
       "658718     22  51217954        A,G    rs9616974    rs9616974  22:51217954:G:A\n",
       "\n",
       "[639250 rows x 6 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_map"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
