unfold_parent_illness_code = function(dd, tbl_code, tbl_name) {
  out = matrix(0, nrow = nrow(dd), ncol = length(tbl_code))
  for(i in 1 : ncol(dd)) {
    dcol = dd[, i]
    for(j in 1 : ncol(out)) {
      has_code = !is.na(dcol) & dcol == tbl_code[j]
      out[has_code, j] = out[has_code, j] + 1
    }
  }
  out = as.data.frame(out)
  colnames(out) = tbl_name
  return(out)
}

# use to clean up the table copy/pasted from data field Data tab
clean_up = function(filename) {
  e = read.delim2(filename)
  write.table(e, filename, sep = ',', col.names = T, row.names = F, quote = F)
}

beta2pval = function(value_beta, value_se) {
  exp(pnorm(-abs(value_beta / value_se), log.p = T)) * 2
}

# to load gwas results generated by logistic_gpu
to_df = function(mat) {
  df = as.data.frame(mat)
  colnames(df) = c('beta', 'se', 'convergence')
  return(df)
}
read_from_npy = function(filename, pos, trait_idx) {
  check_imputer = np$load(filename)
  df_trait2 = rbind(
    to_df(check_imputer[1, trait_idx, , ]) %>% 
      mutate(snpid = 1 : nrow(.), position = pos, impute = 'imp'),
    to_df(check_imputer[2, trait_idx, , ]) %>% 
      mutate(snpid = 1 : nrow(.), position = pos, impute = 'avg'),
    to_df(check_imputer[3, trait_idx, , ]) %>% 
      mutate(snpid = 1 : nrow(.), position = pos, impute = 'flip')
  )
  # extract snps that converge
  df_trait2 = df_trait2 %>% filter(convergence == 1, !is.na(se), se > 0) %>% filter(abs(beta) < 100, se < 5)
  df_trait2
}
signed_beta = function(beta, allele_ids, a1, a2) {
  ref = unlist(lapply(strsplit(allele_ids, ','), function(x) {x[1]}))
  alt = unlist(lapply(strsplit(allele_ids, ','), function(x) {x[2]}))
  obeta = rep(NA, length(beta))
  obeta[ref == a2 & alt == a1] = beta[ref == a2 & alt == a1]
  obeta[ref == a1 & alt == a2] = -beta[ref == a1 & alt == a2]
  return(obeta)
}
get_id = function(id) {
  unlist(lapply(strsplit(id, ':'), function(x) {
    paste0(x[1], ':', x[2])
  }))
}