<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Dive in simulation for haplotype imputation</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Imputing parent of origin of haplotype</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/liangyy/haplotype-po">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Dive in simulation for haplotype imputation</h1>
<h4 class="date">Last update: May 29, 2020</h4>

</div>


<pre class="r"><code>rm(list = ls())
library(dplyr)
library(ggplot2)
theme_set(theme_bw(base_size = 15)) </code></pre>
<div id="about" class="section level1">
<h1><span class="header-section-number">1</span> About</h1>
<p>In this document, we focus on the simulation for haplotype imputation. In particular, we mainly want to address the following questions:</p>
<ol style="list-style-type: decimal">
<li>For heritability at 0.1%, 0.5%, 1% 5% (which is quite typical in real data analysis), how the imputation quality depends on the number of phenotypes.</li>
<li>For the “bad imputation”, what is the source of it and how it depends on the number of phenotypes.</li>
</ol>
<p>For the sake of simplicity, we plug in the true <span class="math inline">\(\beta\)</span> and <span class="math inline">\(\sigma^2\)</span> when evaluating the posterior of <span class="math inline">\(Z\)</span>. In this sense, we are seeking the “best” we could get out of the model and data.</p>
</div>
<div id="simulation-scheme" class="section level1">
<h1><span class="header-section-number">2</span> Simulation scheme</h1>
<p>I will take a slightly different approach as comparing to the previous simulation for the preliminary runs on imputation scheme. Since here we work with the true <span class="math inline">\(\beta\)</span> and <span class="math inline">\(\sigma\)</span>, I skip simulation the genotype and <span class="math inline">\(\beta\)</span> but treat <span class="math inline">\(H\beta\)</span> as a whole. Essentially, the things that we need to simulate are just: 1) genetic effect (for each parent, and they are transmitted to the child); 2) environmental effect (for both parents and child). The <strong>simulation procedure</strong> is as follow:</p>
<ol style="list-style-type: decimal">
<li>Simulate genetic effect, <span class="math inline">\(G^{p,1}\)</span> and <span class="math inline">\(G^{p,2}\)</span>, for each parent of each individual.</li>
<li>Simulate environmental effect, <span class="math inline">\(E^p\)</span>, for each parent of each individual.</li>
<li>Observed phenotype of parents is calculated as <span class="math inline">\(Y^p = G^{p,1} + G^{p,2} + E^p\)</span>.</li>
<li>Simulate the transmission of genetic effect from parents to child for each individual, which gives <span class="math inline">\(G^1\)</span> and <span class="math inline">\(G^2\)</span> for each individuals.</li>
<li>At the end, what we do observe is <span class="math inline">\(Y^f, Y^m, G^1, G^2\)</span>.</li>
</ol>
<p>The followings are some parameters used in the simulation.</p>
<pre class="r"><code>heritability = c(0.001, 0.005, 0.01, 0.05)
nsample = 10000
max_npheno = 100</code></pre>
<ul>
<li>Number of samples (for each sample, we will simulate the father and mother): <span class="math inline">\(10^{4}\)</span>.</li>
<li>Heritability: <span class="math inline">\(h^2 = 0.001, 0.005, 0.01, 0.05\)</span> (we will take phenotype as variance = 1 so that <span class="math inline">\(\sigma^2 = 1 - h^2\)</span>).</li>
<li>Max number of phenotypes: 100</li>
</ul>
</div>
<div id="simulation" class="section level1">
<h1><span class="header-section-number">3</span> Simulation</h1>
<p>This section is just code. I did the very explicit coding to make it as readable as possible. The short notice is that <span class="math inline">\(G^1\)</span> comes from father and <span class="math inline">\(G^2\)</span> comes from mother in this simulation. Skip if you’re comfortable with skipping.</p>
<pre class="r"><code>G1_by_h2 = list()
G2_by_h2 = list()
Yf_by_h2 = list()
Ym_by_h2 = list()

# loop over all heritabilities
for(h2 in heritability) {
  G1_list = list()
  G2_list = list()
  Yf_list = list()
  Ym_list = list()
  
  # loop over phenotypes
  for(pp in 1 : max_npheno) {
    # parental variables
    Gf1 = rnorm(nsample, sd = sqrt(h2 / 2))
    Gf2 = rnorm(nsample, sd = sqrt(h2 / 2))
    Gm1 = rnorm(nsample, sd = sqrt(h2 / 2))
    Gm2 = rnorm(nsample, sd = sqrt(h2 / 2))
    Ef = rnorm(nsample, sd = sqrt(1 - h2))
    Em = rnorm(nsample, sd = sqrt(1 - h2))
    Yf_list[[length(Yf_list) + 1]] = Gf1 + Gf2 + Ef
    Ym_list[[length(Ym_list) + 1]] = Gm1 + Gm2 + Em
    # transmission to child
    G1_list[[length(G1_list) + 1]] = Gf1
    G2_list[[length(G2_list) + 1]] = Gm1
  }
  
  # collecting results
  Yf = do.call(cbind, Yf_list)
  Ym = do.call(cbind, Ym_list)
  G1 = do.call(cbind, G1_list)
  G2 = do.call(cbind, G2_list)
  
  # add to *_by_h2 
  Yf_by_h2[[as.character(h2)]] = Yf
  Ym_by_h2[[as.character(h2)]] = Ym
  G1_by_h2[[as.character(h2)]] = G1
  G2_by_h2[[as.character(h2)]] = G2
}</code></pre>
</div>
<div id="best-imputation" class="section level1">
<h1><span class="header-section-number">4</span> “Best” imputation</h1>
<p>With many simplification due to the setup (knowing true <span class="math inline">\(\beta\)</span> and <span class="math inline">\(\sigma^2\)</span> and both sexes share the same model), the “best” imputation of <span class="math inline">\(Z\)</span> (indicating <span class="math inline">\(G^1\)</span> is from father) is:</p>
<p><span class="math display">\[\begin{aligned}
  \gamma &amp;:= \Pr(Z = 1 | Y^f, Y^m, G^1, G^2) \\
  &amp;= \frac{1}{1 + e^{-\frac{L_0 - L_1}{2\sigma^2}}} \\
  L_1 &amp;:= \sum_j (Y^{f, j} - G^{1, j})^2 + (Y^{m, j} - G^{2, j})^2 \\
  L_0 &amp;:= \sum_j (Y^{f, j} - G^{2, j})^2 + (Y^{m, j} - G^{1, j})^2
\end{aligned}\]</span> where <span class="math inline">\(j\)</span> indicates phenotype.</p>
<pre class="r"><code>slice_m = function(mat, p) {
  mat[, 1 : p, drop = FALSE]
}

sq_diff = function(y, g) {
  rowSums((g - y) ^ 2)
}
eval_gamma = function(yf, ym, g1, g2, sigma2) {
  L1 = sq_diff(yf, g1) + sq_diff(ym, g2)
  L0 = sq_diff(yf, g2) + sq_diff(ym, g1)
  1 / (1 + exp(- (L0 - L1) / 2 / sigma2))
}

# loop over all heritabilities
gamma_list = list()
for(h2 in heritability) {
  Yf = Yf_by_h2[[as.character(h2)]]
  Ym = Ym_by_h2[[as.character(h2)]]
  G1 = G1_by_h2[[as.character(h2)]]
  G2 = G2_by_h2[[as.character(h2)]]
  # loop over the number of phenotypes
  for(pp in 1 : max_npheno) {
    gamma_list[[length(gamma_list) + 1]] = data.frame(
      gamma = eval_gamma(slice_m(Yf, pp), slice_m(Ym, pp), slice_m(G1, pp), slice_m(G2, pp), 1 - h2), 
      npheno = pp,
      heritability = h2
    )
  }
}
gamma_df = do.call(rbind, gamma_list)</code></pre>
<p>We visualize by plotting the <span class="math inline">\(\gamma\)</span>’s at each number of phenotypes. Additionally, we calculate the mean, the variance, and the <span class="math inline">\(S\)</span> quantity at each number of phentoypes.</p>
<pre class="r"><code>gamma_df %&gt;% filter(npheno %in% seq(1,100, by = 5)) %&gt;% ggplot() + geom_jitter(aes(x = npheno, y = gamma), alpha = 0.05, height = 0, width = 0.5) + facet_wrap(~heritability)</code></pre>
<p><img src="dive_in_simulation_files/figure-html/best%20imputation%20vis%20scatter-1.png" width="1440" /></p>
<pre class="r"><code>gamma_df %&gt;% filter(npheno %in% seq(1,100, by = 5)) %&gt;% ggplot() + 
  geom_violin(aes(x = npheno, y = gamma, group = npheno), scale = &#39;width&#39;) +
  geom_boxplot(aes(x = npheno, y = gamma, group = npheno)) +
  facet_wrap(~heritability)</code></pre>
<p><img src="dive_in_simulation_files/figure-html/best%20imputation%20vis%20violin-1.png" width="1440" /></p>
<pre class="r"><code>get_sval = function(gamma) {
  mean(gamma ^ 2 + (1 - gamma) ^ 2)
}
gamma_df %&gt;% group_by(npheno, heritability) %&gt;% 
  summarise(mean_gamma = mean(gamma), var_gamma = var(gamma), sval = get_sval(gamma)) %&gt;% 
  ungroup() %&gt;% mutate(power_against_avg = mean_gamma / sqrt(sval) * sqrt(2), power_against_optimal = mean_gamma / sqrt(sval)) %&gt;%
  ggplot() + geom_path(aes(x = npheno, y = mean_gamma, color = &#39;mean_gamma&#39;)) + 
  geom_path(aes(x = npheno, y = var_gamma, color = &#39;var_gamma&#39;)) +
  geom_path(aes(x = npheno, y = sval, color = &#39;sval&#39;)) + 
  geom_path(aes(x = npheno, y = power_against_avg, color = &#39;power_against_avg&#39;)) + 
  geom_path(aes(x = npheno, y = power_against_optimal, color = &#39;power_against_optimal&#39;)) +
  facet_wrap(~heritability)</code></pre>
<p><img src="dive_in_simulation_files/figure-html/best%20imputation%20curve-1.png" width="672" /></p>
<p><strong>One not so important but good to know observation</strong> is that <code>sval</code> is very close to <code>mean_gamma</code>. The difference between <span class="math inline">\(S\)</span> and <span class="math inline">\(\gamma^\star\)</span> is <span class="math inline">\(\delta = 2 (\gamma - 0.5) (\gamma - 1)\)</span> which is a quadratic function (here I use <span class="math inline">\(\star\)</span> to denote that the <span class="math inline">\(\gamma\)</span> is the probability of correct but not the imputation output). If we plot this function, we will see that it is close to zero when <span class="math inline">\(\gamma^\star \in [0.5, 1]\)</span>.</p>
<pre class="r"><code>diff_sval_mean_gamma = function(g) {
  2 * (g - 0.5) * (g - 1) 
}
plot(seq(0, 1, by = 0.01),diff_sval_mean_gamma(seq(0, 1, by = 0.01)))</code></pre>
<p><img src="dive_in_simulation_files/figure-html/vis%20diff%20sval%20vs%20gamma-1.png" width="672" /></p>
<p>This means that if the imputation is not so off, in the sense that <span class="math inline">\(\gamma^\star \in [0.5, 1]\)</span>, <span class="math inline">\(\bar{S}\)</span> should be close to <span class="math inline">\(\bar{\gamma^\star}\)</span>.</p>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
