<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Imputation results of Framingham data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Imputing parent of origin of haplotype</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/liangyy/haplotype-po">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Imputation results of Framingham data</h1>
<h4 class="date">Last update: Jun 11, 2020</h4>

</div>


<p><span class="math display">\[
\newcommand{\E}{\text{E}}
\newcommand{\diag}{\text{diag}}
\]</span></p>
<pre class="r"><code>rm(list = ls())
library(dplyr)
library(ggplot2)
library(data.table)
theme_set(theme_bw(base_size=15))  
bin_prob_z = function(x, cuts = c(0.1, 0.4, 0.6, 0.9)) {
  left_most = cuts[1]
  left_mid = cuts[2]
  right_mid = cuts[3]
  right_most = cuts[4]
  o = rep(&#39;&#39;, length(x))
  o[x &lt; left_most] = &#39;strong error&#39;
  o[x &gt;= left_most &amp; x &lt; left_mid] = &#39;mild error&#39;
  o[x &gt;= left_mid &amp; x &lt;= right_mid] = &#39;not sure&#39;
  o[x &gt; right_mid &amp; x &lt;= right_most] = &#39;mild correct&#39;
  o[x &gt; right_most] = &#39;strong correct&#39;
  o = factor(o, levels = c(&#39;strong error&#39;, &#39;mild error&#39;, &#39;not sure&#39;, &#39;mild correct&#39;, &#39;strong correct&#39;))
  o
}
get_ngene = function(logfile) {
  e = fread(cmd = paste0(&#39;cat &#39;, logfile, &#39; | grep SUMMARY&#39;), sep = &#39; &#39;, data.table = FALSE)
  e = e[length(e) - 3]
  as.numeric(e)
}
power_calc = function(gamma) {
  mean_z = mean(gamma)
  mean_s = mean( gamma ^ 2 + (1 - gamma) ^ 2)
  mean_z / sqrt(mean_s)
}</code></pre>
<div id="about" class="section level1">
<h1><span class="header-section-number">1</span> About</h1>
<p>The effort to analyze Framingham data is motivated by the failure on UK Biobank data where we have a tiny set of parental phenotypes to start with. Also, simulation results indicate that the more parental phenotypes the greater imputation power would be. So, we decided to dig into Framingham data where we have some child/father/mother tuple with both transcriptome and genotype available.</p>
<p>The code for this task is at <code>scripts/framingham_detour/</code>.</p>
</div>
<div id="haplotype-origin-inferred-from-genotypes-in-family" class="section level1">
<h1><span class="header-section-number">2</span> Haplotype origin inferred from genotypes in family</h1>
<pre class="r"><code>chroms = 1 : 22
df_from_geno = list()
for(chrom in chroms) {
  df = fread(cmd = paste0(&#39;zcat &lt; ~/Desktop/tmp/haplotype-po/framingham/infer_ho_from_genotype/from_extracted_pedigree.subset.chr&#39;, chrom, &#39;.ho_from_genotype.tsv.gz&#39;), header = TRUE, sep = &#39;\t&#39;, data.table = FALSE)
  df_from_geno[[length(df_from_geno) + 1]] = df %&gt;% mutate(chr = chrom)
}
df_from_geno = do.call(rbind, df_from_geno)</code></pre>
<pre class="r"><code>df_from_geno %&gt;% summarize(mean(prob_z))</code></pre>
<pre><code>##   mean(prob_z)
## 1            1</code></pre>
<p>So, the crude approach of assigning haplotype origin indicates that haplotype 1 of child is from father for all pedigrees.</p>
<p>To further verify the result, we look at the observed kinship. The observed kinship between child’s haplotype and parents’ genotypes, which is defined as follow with standardized haplotype.</p>
<ul>
<li>For diploid (father and mother): <span class="math inline">\(\widetilde{H} = \frac{H^1}{\sqrt{2}} + \frac{H^2}{\sqrt{2}}\)</span> (So that <span class="math inline">\(\widetilde{H}\)</span> has mean 0 and variance 1)</li>
<li>For haploid (child), <span class="math inline">\(\widetilde{H} = H\)</span>.</li>
<li>Observed kinship is defined as
<ul>
<li>Between child haplotype and parent genotype: <span class="math inline">\(K_{1,p} = \frac{1}{K} \sum_k H_k^1 \widetilde{H}_k^p\)</span></li>
<li>Between parents: <span class="math inline">\(K_{p, q} = \frac{1}{K} \sum_i \widetilde{H}_k^p \widetilde{H}_k^q\)</span></li>
</ul></li>
<li>So that, the expected value of kinship is</li>
</ul>
<p><span class="math display">\[\begin{aligned}
  \E(K_{1, p}) &amp;= \frac{1}{K} \sum_k \frac{1}{\sqrt{2}}[~ \E(H_k^1 H_k^{p,1}) + \E(H_k^1 H_k^{p,2}) ~] \\
  &amp;= \frac{1}{\sqrt{2}} \\
  \E(K_{p, q}) &amp;= \begin{cases}
    \frac{1}{K} \sum_k \frac{1}{2} [~ \E((H_k^{p, 1})^2) + \E((H_k^2{p, 1})^2) ~] = 1 &amp; \text{, if $p = q$} \\
    0 &amp; \text{, otherwise}
  \end{cases}
\end{aligned}\]</span></p>
<pre class="r"><code>df_geno_cor = list()
for(chrom in chroms) {
  df = fread(cmd = paste0(&#39;zcat &lt; ~/Desktop/tmp/haplotype-po/framingham/infer_ho_from_genotype/from_extracted_pedigree.subset.chr&#39;, chrom, &#39;.child_parents_cor.tsv.gz&#39;), header = TRUE, sep = &#39;\t&#39;, data.table = FALSE)
  df_geno_cor[[length(df_geno_cor) + 1]] = df %&gt;% mutate(chr = chrom)
}
df_geno_cor = do.call(rbind, df_geno_cor)</code></pre>
<pre class="r"><code>df_geno_cor %&gt;% reshape2::melt(id.vars = c(&#39;individual_id&#39;, &#39;chr&#39;)) %&gt;% ggplot() + geom_boxplot(aes(x = variable, y = value)) + facet_wrap(~chr) + theme(axis.text.x = element_text(angle = 90)) + geom_hline(yintercept = 1 / sqrt(2), linetype = 2, color = &#39;gray&#39;)</code></pre>
<p><img src="framingham_results_files/figure-html/unnamed-chunk-27-1.png" width="960" /></p>
<pre class="r"><code># ggsave(&#39;~/Desktop/temp.png&#39;, p, width = 10, height = 10)</code></pre>
<p>So, we have high confidence to say that “haplotype 1 of child comes from father for all families”.</p>
</div>
<div id="load-results" class="section level1">
<h1><span class="header-section-number">3</span> Load results</h1>
<p>Imputation on full data.</p>
<pre class="r"><code>models = c(&#39;en&#39;, &#39;dapgw&#39;)
df_impute = list()
df_ngene = list()
for(chrom in chroms) {
  for(mod in models) {
    df = fread(cmd = paste0(&#39;zcat &lt; ~/Desktop/tmp/haplotype-po/framingham/imputation/haplo_impute.&#39;, mod, &#39;.chr&#39;, chrom, &#39;.tsv.gz&#39;), header = TRUE, sep = &#39;\t&#39;, data.table = FALSE)
    logfile = paste0(&#39;~/Desktop/tmp/haplotype-po/framingham/imputation/haplo_impute.&#39;, mod, &#39;.chr&#39;, chrom, &#39;.log&#39;)
    df_impute[[length(df_impute) + 1]] = df %&gt;% mutate(chr = chrom, model = mod, fraction = 1)
    df_ngene[[length(df_ngene) + 1]] = data.frame(chr = chrom, model = mod, fraction = 1, ngene = get_ngene(logfile))
  }
}
# df_impute = do.call(rbind, df_impute)</code></pre>
<p>Imputation on downsampled data.</p>
<pre class="r"><code>fractions = 1 : 9 / 10
# df_impute = list()
for(chrom in chroms) {
  for(mod in models) {
    for(frac in fractions) {
      df = fread(cmd = paste0(&#39;zcat &lt; ~/Desktop/tmp/haplotype-po/framingham/imputation/haplo_impute.downsample_&#39;, frac, &#39;.&#39;, mod, &#39;.chr&#39;, chrom, &#39;.tsv.gz&#39;), header = TRUE, sep = &#39;\t&#39;, data.table = FALSE)
      logfile = paste0(&#39;~/Desktop/tmp/haplotype-po/framingham/imputation/haplo_impute.downsample_&#39;, frac, &#39;.&#39;, mod, &#39;.chr&#39;, chrom, &#39;.log&#39;)
      df_impute[[length(df_impute) + 1]] = df %&gt;% mutate(chr = chrom, model = mod, fraction = frac) 
      df_ngene[[length(df_ngene) + 1]] = data.frame(chr = chrom, model = mod, fraction = frac, ngene = get_ngene(logfile))
    }
  }
}
df_impute = do.call(rbind, df_impute)
df_impute = df_impute %&gt;% mutate(binned = bin_prob_z(prob_z))
df_ngene = do.call(rbind, df_ngene) %&gt;% mutate(label = paste0(&#39;chr&#39;, chr, &#39; &#39;, model, &#39;\n ngene = &#39;, ngene)) %&gt;% mutate(label = factor(label, levels = label[order(chr)]))</code></pre>
<div id="gamma-from-imputation" class="section level2">
<h2><span class="header-section-number">3.1</span> <span class="math inline">\(\gamma\)</span> from imputation</h2>
<pre class="r"><code>df_impute %&gt;% filter(fraction == 1) %&gt;% left_join(df_ngene %&gt;% filter(fraction == 1), by = c(&#39;chr&#39;, &#39;model&#39;, &#39;fraction&#39;)) %&gt;% filter(model == &#39;en&#39;, fraction == 1) %&gt;% ggplot() + geom_histogram(aes(prob_z)) + facet_wrap(~label) + ggtitle(&#39;EN: distribution of gamma&#39;)</code></pre>
<pre><code>## Warning: Column `model` joining character vector and factor, coercing into character vector</code></pre>
<pre><code>## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="framingham_results_files/figure-html/unnamed-chunk-30-1.png" width="960" /></p>
<pre class="r"><code>df_impute %&gt;% filter(fraction == 1) %&gt;% left_join(df_ngene %&gt;% filter(fraction == 1), by = c(&#39;chr&#39;, &#39;model&#39;, &#39;fraction&#39;)) %&gt;% filter(model == &#39;dapgw&#39;, fraction == 1) %&gt;% ggplot() + geom_histogram(aes(prob_z)) + facet_wrap(~label) + ggtitle(&#39;DAPGW: distribution of gamma&#39;)</code></pre>
<pre><code>## Warning: Column `model` joining character vector and factor, coercing into character vector</code></pre>
<pre><code>## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="framingham_results_files/figure-html/unnamed-chunk-31-1.png" width="960" /></p>
</div>
<div id="downsample-analysis" class="section level2">
<h2><span class="header-section-number">3.2</span> Downsample analysis</h2>
<p>Downsample the number of genes by keeping a fraction of the total genes randomly.</p>
<pre class="r"><code>df_impute %&gt;% filter(model == &#39;en&#39;) %&gt;% left_join(df_ngene %&gt;% filter(fraction == 1) %&gt;% select(-fraction), by = c(&#39;chr&#39;, &#39;model&#39;)) %&gt;% ggplot() + geom_bar(aes(binned, fill = fraction, group = fraction), position = &#39;dodge&#39;) + facet_wrap(~label) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) + ggtitle(&#39;EN: Downsample&#39;)</code></pre>
<pre><code>## Warning: Column `model` joining character vector and factor, coercing into character vector</code></pre>
<p><img src="framingham_results_files/figure-html/unnamed-chunk-32-1.png" width="960" /></p>
<pre class="r"><code>df_impute %&gt;% filter(model == &#39;dapgw&#39;) %&gt;% left_join(df_ngene %&gt;% filter(fraction == 1) %&gt;% select(-fraction), by = c(&#39;chr&#39;, &#39;model&#39;)) %&gt;% ggplot() + geom_bar(aes(binned, fill = fraction, group = fraction), position = &#39;dodge&#39;) + facet_wrap(~label) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) + ggtitle(&#39;DAPGW: Downsample&#39;)</code></pre>
<pre><code>## Warning: Column `model` joining character vector and factor, coercing into character vector</code></pre>
<p><img src="framingham_results_files/figure-html/unnamed-chunk-33-1.png" width="960" /></p>
</div>
<div id="power-of-gwas-with-imputed-haplotype" class="section level2">
<h2><span class="header-section-number">3.3</span> Power of GWAS with imputed haplotype</h2>
<pre class="r"><code>df_impute %&gt;% group_by(model, chr, fraction) %&gt;% summarize(power = power_calc(prob_z)) %&gt;% ungroup() %&gt;% left_join(df_ngene, by = c(&#39;chr&#39;, &#39;model&#39;, &#39;fraction&#39;)) %&gt;% mutate(power_gain_relative_to_avg = power * sqrt(2)) %&gt;% ggplot() + geom_point(aes(x = ngene, y = power_gain_relative_to_avg, color = model)) + facet_wrap(~chr) + geom_hline(yintercept = 1, linetype = 2, color = &#39;gray&#39;) + geom_vline(xintercept = 8, linetype = 2, color = &#39;red&#39;) + scale_x_log10()</code></pre>
<pre><code>## Warning: Column `model` joining character vector and factor, coercing into character vector</code></pre>
<p><img src="framingham_results_files/figure-html/unnamed-chunk-34-1.png" width="960" /></p>
<pre class="r"><code># ggsave(&#39;temp.png&#39;, p, width = 10, height = 10)</code></pre>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
