<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Genome-wide run (multi-chromosome imputation)</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Imputing parent of origin of haplotype</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/liangyy/haplotype-po">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Genome-wide run (multi-chromosome imputation)</h1>
<h4 class="date">Last update: May 30, 2020</h4>

</div>


<pre class="r"><code>rm(list = ls())
library(ggplot2)
library(dplyr)
library(patchwork)
library(data.table)
library(reticulate)
library(pander)
options(datatable.fread.datatable = FALSE)
np &lt;- import(&quot;numpy&quot;)
set.seed(2020)
source(&#39;../code/rlib_misc.R&#39;)</code></pre>
<div id="about" class="section level1">
<h1><span class="header-section-number">1</span> About</h1>
<p>I’ve implemented a pipeline doing the haplotype imputation based GWAS which includes three major scripts:</p>
<ul>
<li><strong>Calculate PRS</strong>: <code>scripts/prs/naive_prs.py</code></li>
<li><strong>Imputation</strong>: <code>scripts/haplotype_imputation/impute_parent_of_origin.py</code></li>
<li><strong>Run GWAS</strong>: <code>scripts/logistic_gpu/run_haplo_logistic_solver.py</code></li>
</ul>
<p>As first pass effort, I run GWAS on paternal phenotype with imputed haplotype (Alzheimer’s disease, AD). The result is directly comparable to the paternal GWAS on AD reported in <span class="citation">Marioni et al. (2018)</span>. I tested on chromosome 16 and 19.</p>
<p>Two imputation (labeled as <code>imp</code>) approaches are tested: 1) sex-specific EM (labeled as <code>em</code>); 2) non sex-specific EM (as <code>em_deg</code>). Besides, I include <span class="math inline">\(\Pr(Z) = 0.5\)</span> as <code>avg</code>, and <code>flip</code> which flip the imputation by using <span class="math inline">\(1 - Pr(Z)\)</span>.</p>
<p>Besides, I also run on heart disease (HD) on the side, though I don’t have an existing GWAS to compare to.</p>
</div>
<div id="load-results" class="section level1">
<h1><span class="header-section-number">2</span> Load results</h1>
<pre class="r"><code>snp_map = fread(&#39;zcat &lt; ~/Desktop/tmp/haplotype-po/snp_map_for_neale_lab_gwas.with_sign.tsv.gz&#39;, header = TRUE, sep = &#39;\t&#39;)

dl = list()
dl2 = list()
chromosomes = 1 : 22
traits = c(&#39;HD&#39;, &#39;AD&#39;, &#39;HT&#39;)
for(chr_num in chromosomes) {
  pos = snp_map %&gt;% filter(chrom == chr_num) %&gt;% pull(pos)
  for(i in 1 : length(traits)) {
    filename = paste0(&#39;/Users/yanyul/Desktop/tmp/haplotype-po/from_lambda/gwas_multi_chr/chr&#39;, chr_num, &#39;.npy&#39;)
    df = read_from_npy(
      paste0(filename),
      pos, 
      i
    )
    df = df %&gt;% mutate(chr = chr_num, trait = traits[i])
    dl[[length(dl) + 1]] = df
  }
  for(i in 1 : length(traits)) {
    filename = paste0(&#39;/Users/yanyul/Desktop/tmp/haplotype-po/from_lambda/gwas/chr&#39;, chr_num, &#39;.npy&#39;)
    df = read_from_npy(
      paste0(filename),
      pos, 
      i
    )
    df = df %&gt;% mutate(chr = chr_num, trait = traits[i])
    dl2[[length(dl2) + 1]] = df
  }
}
df = do.call(rbind, dl)
df2 = do.call(rbind, dl2)
df = rbind(df %&gt;% mutate(impute_scheme = &#39;all_chr&#39;), df2 %&gt;% mutate(impute_scheme = &#39;per_chr&#39;))</code></pre>
</div>
<div id="ht" class="section level1">
<h1><span class="header-section-number">3</span> HT</h1>
<pre class="r"><code>df_ht = df %&gt;% filter(trait == &#39;HT&#39;) %&gt;% mutate(pval = beta2pval(beta, se))</code></pre>
<div id="neale-lab-gwas" class="section level2">
<h2><span class="header-section-number">3.1</span> Neale lab GWAS</h2>
<pre class="r"><code>gwas_self = fread(&#39;zcat &lt; ~/Desktop/tmp/haplotype-po/20002_1065.gwas.imputed_v3.both_sexes.tsv.bgz|awk \&#39;{split($1,a,&quot;:&quot;); print $0&quot;\t&quot;a[1]&quot;:&quot;a[2]}\&#39;&#39;, header = T, sep = &#39;\t&#39;, data.table = F)
# gwas_parent = fread(&#39;zcat &lt; ~/Desktop/tmp/haplotype-po/20107_8.gwas.imputed_v3.both_sexes.tsv.bgz | grep &quot;^16:\\|^19:&quot;&#39;, header = F, sep = &#39;\t&#39;, data.table = F)
gwas_self = gwas_self %&gt;% mutate(snp = `variant:`)
gwas_self = gwas_self %&gt;% filter(snp %in% unique(paste0(df_ht$chr, &#39;:&#39;, df_ht$position)))

gwas_ht_bonferroni = gwas_self %&gt;% filter(pval &lt; 0.1 / n(), !low_confidence_variant)</code></pre>
<pre class="r"><code>df_ht %&gt;% filter(paste0(chr, &#39;:&#39;, position) %in% gwas_ht_bonferroni$snp) %&gt;% reshape2::dcast(position + chr + impute_scheme ~ impute, value.var = &#39;pval&#39;) %&gt;% ggplot() + geom_point(aes(x = -log(avg), y = -log(flip), color = &#39;flip&#39;), alpha = .3) + geom_point(aes(x = -log(avg), y = -log(imp), color = &#39;imp&#39;), alpha = .3) + geom_abline(slope = 1, intercept = 0) + coord_equal() + facet_wrap(~impute_scheme)</code></pre>
<p><img src="genomewide_check_multi_chr_files/figure-html/unnamed-chunk-69-1.png" width="672" /></p>
<pre class="r"><code>df_ht %&gt;% filter(paste0(chr, &#39;:&#39;, position) %in% gwas_ht_bonferroni$snp) %&gt;% reshape2::dcast(position + chr + impute_scheme ~ impute, value.var = &#39;pval&#39;) %&gt;% ggplot() + geom_point(aes(x = -log(flip), y = -log(imp))) + geom_density2d(aes(x = -log(flip), y = -log(imp))) + geom_abline(slope = 1, intercept = 0, color = &#39;gray&#39;) + coord_equal() + facet_wrap(~impute_scheme)</code></pre>
<p><img src="genomewide_check_multi_chr_files/figure-html/unnamed-chunk-69-2.png" width="672" /></p>
<pre class="r"><code>df_ht %&gt;% filter(paste0(chr, &#39;:&#39;, position) %in% gwas_ht_bonferroni$snp) %&gt;% reshape2::dcast(position + chr + impute_scheme ~ impute, value.var = &#39;pval&#39;) %&gt;% ggplot() + geom_point(aes(x = -log(avg), y = -log(imp))) + geom_density2d(aes(x = -log(avg), y = -log(imp))) + geom_abline(slope = 1, intercept = 0, color = &#39;gray&#39;) + coord_equal() + facet_wrap(~impute_scheme)</code></pre>
<p><img src="genomewide_check_multi_chr_files/figure-html/unnamed-chunk-69-3.png" width="672" /></p>
<pre class="r"><code>df_ht %&gt;% filter(paste0(chr, &#39;:&#39;, position) %in% gwas_ht_bonferroni$snp) %&gt;% filter(chr %in% c(&#39;21&#39;, &#39;22&#39;), impute == &#39;imp&#39;) %&gt;% reshape2::dcast(position + chr ~ impute_scheme, value.var = &#39;pval&#39;) %&gt;% ggplot() + geom_point(aes(x = -log(per_chr), y = -log(all_chr))) + geom_abline(slope = 1, intercept = 0, color = &#39;gray&#39;) + coord_equal() + facet_wrap(~chr)</code></pre>
<p><img src="genomewide_check_multi_chr_files/figure-html/unnamed-chunk-70-1.png" width="672" /></p>
<pre class="r"><code>df_ht  %&gt;% filter(paste0(chr, &#39;:&#39;, position) %in% gwas_ht_bonferroni$snp) %&gt;% filter(chr %in% c(&#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;), impute == &#39;imp&#39;) %&gt;% reshape2::dcast(position + chr ~ impute_scheme, value.var = &#39;pval&#39;) %&gt;% ggplot() + geom_point(aes(x = -log(per_chr), y = -log(all_chr))) + geom_abline(slope = 1, intercept = 0, color = &#39;gray&#39;) + coord_equal() + facet_wrap(~chr)</code></pre>
<p><img src="genomewide_check_multi_chr_files/figure-html/unnamed-chunk-71-1.png" width="672" /></p>
<pre class="r"><code>df_ht  %&gt;% filter(paste0(chr, &#39;:&#39;, position) %in% gwas_ht_bonferroni$snp) %&gt;% filter(chr %in% c(&#39;6&#39;, &#39;7&#39;, &#39;8&#39;, &#39;9&#39;, &#39;10&#39;), impute == &#39;imp&#39;) %&gt;% reshape2::dcast(position + chr ~ impute_scheme, value.var = &#39;pval&#39;) %&gt;% ggplot() + geom_point(aes(x = -log(per_chr), y = -log(all_chr))) + geom_abline(slope = 1, intercept = 0, color = &#39;gray&#39;) + coord_equal() + facet_wrap(~chr)</code></pre>
<p><img src="genomewide_check_multi_chr_files/figure-html/unnamed-chunk-72-1.png" width="672" /></p>
<pre class="r"><code>df_ht  %&gt;% filter(paste0(chr, &#39;:&#39;, position) %in% gwas_ht_bonferroni$snp) %&gt;% filter(chr %in% c(&#39;11&#39;, &#39;12&#39;, &#39;13&#39;, &#39;14&#39;, &#39;15&#39;), impute == &#39;imp&#39;) %&gt;% reshape2::dcast(position + chr ~ impute_scheme, value.var = &#39;pval&#39;) %&gt;% ggplot() + geom_point(aes(x = -log(per_chr), y = -log(all_chr))) + geom_abline(slope = 1, intercept = 0, color = &#39;gray&#39;) + coord_equal() + facet_wrap(~chr)</code></pre>
<p><img src="genomewide_check_multi_chr_files/figure-html/unnamed-chunk-73-1.png" width="672" /></p>
<pre class="r"><code>df_ht  %&gt;% filter(paste0(chr, &#39;:&#39;, position) %in% gwas_ht_bonferroni$snp) %&gt;% filter(chr %in% c(&#39;16&#39;, &#39;17&#39;, &#39;18&#39;, &#39;19&#39;, &#39;20&#39;), impute == &#39;imp&#39;) %&gt;% reshape2::dcast(position + chr ~ impute_scheme, value.var = &#39;pval&#39;) %&gt;% ggplot() + geom_point(aes(x = -log(per_chr), y = -log(all_chr))) + geom_abline(slope = 1, intercept = 0, color = &#39;gray&#39;) + coord_equal() + facet_wrap(~chr)</code></pre>
<p><img src="genomewide_check_multi_chr_files/figure-html/unnamed-chunk-74-1.png" width="672" /></p>
</div>
<div id="power-gain" class="section level2">
<h2><span class="header-section-number">3.2</span> Power gain</h2>
<p>Evaluate as the median of ratio of <span class="math inline">\(\log(p)\)</span> among genome-wide significant SNPs (Bonferroni correction: 0.1 / nsnp) and the SNPs should have pval &lt; 0.1 in both avg and imp. Another measure: ratio of abs z-score.</p>
<pre class="r"><code>tmp = df_ht %&gt;% filter(paste0(chr, &#39;:&#39;, position) %in% gwas_ht_bonferroni$snp, impute_scheme == &#39;all_chr&#39;) %&gt;% reshape2::dcast(position + chr ~ impute, value.var = &#39;pval&#39;)
tmp %&gt;% filter(imp &lt; 0.1, avg &lt; 0.1) %&gt;% mutate(ratio_logp = log(imp) / log(avg), z_imp = qnorm(imp, lower.tail = F), z_avg = qnorm(avg, lower.tail = F)) %&gt;% mutate(ratio_z = z_imp / z_avg) %&gt;% group_by(chr) %&gt;% summarize(median_ratio_logp = median(ratio_logp), mean_ratio_logp = mean(ratio_logp), nsnp = n(), median_ratio_z = median(ratio_z), mean_ratio_z = mean(ratio_z)) %&gt;% write.csv(&#39;../analysis_output/gw_check_multi_chr.power.csv&#39;, row.names = F)</code></pre>
</div>
<div id="compare-with-per-chromosome-fit-in-terms-of-power" class="section level2">
<h2><span class="header-section-number">3.3</span> Compare with per-chromosome fit (in terms of power)</h2>
<pre class="r"><code>d1 = read.csv(&#39;../analysis_output/gw_check_multi_chr.power.csv&#39;)
d2 = read.csv(&#39;../analysis_output/gw_check.power.csv&#39;)
plot(d1$mean_ratio_z, d2$mean_ratio_z); abline(a = 0, b = 1)</code></pre>
<p><img src="genomewide_check_multi_chr_files/figure-html/unnamed-chunk-76-1.png" width="672" /></p>
</div>
</div>
<div id="ad" class="section level1">
<h1><span class="header-section-number">4</span> AD</h1>
<pre class="r"><code>df_ad = df %&gt;% filter(trait == &#39;AD&#39;) %&gt;% mutate(pval = beta2pval(beta, se))</code></pre>
<pre class="r"><code>gwas_ad = fread(&#39;cat ~/Downloads/4_UKB_IGAP_AD_meta_summary_output_June2019.txt|awk \&#39;{print $0&quot; &quot;$1&quot;:&quot;$2}\&#39;&#39;, header = T, sep = &#39; &#39;, data.table = F)
gwas_ad = gwas_ad %&gt;% mutate(snp = `CHR:BP`)
gwas_ad = gwas_ad %&gt;% filter(snp %in% unique(paste0(df_ht$chr, &#39;:&#39;, df_ht$position)))

gwas_ad_bonferroni = gwas_ad %&gt;% filter(P &lt; 0.1 / n())</code></pre>
<pre class="r"><code>gwas_ad %&gt;% filter(CHR != 19) %&gt;% mutate(pexp = rank(P) / (n() + 1)) %&gt;% ggplot() + geom_point(aes(x = -log(pexp), y = -log(P))) + ggtitle(&#39;QQ-plot of external AD GWAS: \n meta analysis of UKB parental AD + IGAP (exclude chr19)&#39;)</code></pre>
<p><img src="genomewide_check_multi_chr_files/figure-html/unnamed-chunk-78-1.png" width="672" /></p>
<pre class="r"><code>df_ad %&gt;% filter(chr != 19) %&gt;% filter(paste0(chr, &#39;:&#39;, position) %in% gwas_ad_bonferroni$snp) %&gt;% reshape2::dcast(position + chr + impute_scheme ~ impute, value.var = &#39;pval&#39;) %&gt;% ggplot() + geom_point(aes(x = -log(avg), y = -log(flip), color = &#39;flip&#39;), alpha = .3) + geom_point(aes(x = -log(avg), y = -log(imp), color = &#39;imp&#39;), alpha = .3) + geom_abline(slope = 1, intercept = 0) + coord_equal(xlim = c(0, 25), ylim = c(0, 25)) + facet_wrap(~impute_scheme) </code></pre>
<p><img src="genomewide_check_multi_chr_files/figure-html/unnamed-chunk-79-1.png" width="672" /></p>
<pre class="r"><code>df_ad %&gt;% filter(chr != 19) %&gt;% filter(paste0(chr, &#39;:&#39;, position) %in% gwas_ad_bonferroni$snp) %&gt;% reshape2::dcast(position + chr + impute_scheme ~ impute, value.var = &#39;pval&#39;) %&gt;% ggplot() + geom_point(aes(x = -log(flip), y = -log(imp))) + geom_density2d(aes(x = -log(flip), y = -log(imp))) + geom_abline(slope = 1, intercept = 0, color = &#39;gray&#39;) + coord_equal() + facet_wrap(~impute_scheme) + lims(x = c(0, 25), y = c(0, 25))</code></pre>
<p><img src="genomewide_check_multi_chr_files/figure-html/unnamed-chunk-79-2.png" width="672" /></p>
<pre class="r"><code>df_ad %&gt;% filter(chr != 19) %&gt;% filter(paste0(chr, &#39;:&#39;, position) %in% gwas_ad_bonferroni$snp) %&gt;% reshape2::dcast(position + chr + impute_scheme ~ impute, value.var = &#39;pval&#39;) %&gt;% ggplot() + geom_point(aes(x = -log(avg), y = -log(imp))) + geom_density2d(aes(x = -log(avg), y = -log(imp))) + geom_abline(slope = 1, intercept = 0, color = &#39;gray&#39;)  + coord_equal() + facet_wrap(~impute_scheme) + lims(x = c(0, 25), y = c(0, 25))</code></pre>
<p><img src="genomewide_check_multi_chr_files/figure-html/unnamed-chunk-79-3.png" width="672" /></p>
</div>
<div id="references" class="section level1 unnumbered">
<h1>References</h1>
<div id="refs" class="references">
<div id="ref-marioni:2018">
<p>Marioni, Riccardo E, Sarah E Harris, Qian Zhang, Allan F McRae, Saskia P Hagenaars, W David Hill, Gail Davies, et al. 2018. “GWAS on Family History of Alzheimer’s Disease.” <em>Translational Psychiatry</em> 8 (1). Nature Publishing Group: 1–7.</p>
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
