<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Impute haplotype and perform GWAS: power analysis</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Imputing parent of origin of haplotype</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/liangyy/haplotype-po">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Impute haplotype and perform GWAS: power analysis</h1>
<h4 class="date">Last update: May 04, 2020</h4>

</div>


<p><span class="math display">\[
\newcommand{\E}{\text{E}}
\]</span></p>
<pre class="r"><code>rm(list = ls())
library(ggplot2)
library(dplyr)</code></pre>
<pre><code>## 
## Attaching package: &#39;dplyr&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     intersect, setdiff, setequal, union</code></pre>
<pre class="r"><code>theme_set(theme_bw(base_size=15))
set.seed(2020)

# load some gists
source(&#39;https://gist.githubusercontent.com/liangyy/43912b3ecab5d10c89f9d4b2669871c9/raw/3ca651cfa53ffccb8422f432561138a46e93710f/my_ggplot_theme.R&#39;)
source(&#39;https://gist.githubusercontent.com/liangyy/e580a36154586148cca7fd4cd973f502/raw/bad4364b1700662c7086fcdea191e42f530d0e2e/zval2pval.R&#39;)</code></pre>
<div id="about" class="section level1">
<h1><span class="header-section-number">1</span> About</h1>
<p>Here we perform simulation and conduct the full analysis including both imputationa and association test. The goal is to provide a more realistic <span class="math inline">\(\Pr(Z)\)</span> for GWAS analysis (with high variation). At such high variation of <span class="math inline">\(\Pr(Z)\)</span> (some imputations are good but some are very bad), would there still be power gain?</p>
</div>
<div id="simulation" class="section level1">
<h1><span class="header-section-number">2</span> Simulation</h1>
<div id="simulation-for-imputation" class="section level2">
<h2><span class="header-section-number">2.1</span> Simulation for imputation</h2>
<ol style="list-style-type: decimal">
<li>Simulate haplotypes for the parents, <span class="math inline">\(H_i^{f, 1}, H_i^{f, 2}\)</span> and <span class="math inline">\(H_i^{m, 1}, H_i^{m, 2}\)</span>.</li>
<li>Simulate haplotypes for the child, <span class="math inline">\(H_i^1, H_i^2\)</span>.</li>
<li>Simulate phenotypes for the parents, <span class="math inline">\(y_{i, p}^f\)</span> and <span class="math inline">\(y_{i, p}^m\)</span>.</li>
<li>Simulate extra cohort for building PRS, <span class="math inline">\(y_{i, p}\)</span> and <span class="math inline">\(X_{i, p}\)</span></li>
</ol>
<p>Some parameter settings:</p>
<pre class="r"><code>step1_sample_size = 10000
step1_sample_size_extra = 20000
step1_n_pheno = 30
step1_prior_causal = 0.1
step1_causal_sigma = 1
step1_n_snp = 1000
step1_heritabiltiy = 0.05
step1_maf_low = 0.05
step1_maf_high = 0.45</code></pre>
<ul>
<li>Sample size <span class="math inline">\(10^{4}\)</span>.</li>
<li>Sample size for PRS <span class="math inline">\(2\times 10^{4}\)</span>.</li>
<li>Number of phenotypes <span class="math inline">\(30\)</span>.</li>
<li>Effect size <span class="math inline">\(\beta_{p, k} \sim \pi \delta_0 + (1 - \pi) N(0, \sigma)\)</span> where <span class="math inline">\(\pi = 0.9\)</span> and <span class="math inline">\(\sigma^2 = 1\)</span>.</li>
<li>Allele frequency <span class="math inline">\(maf \sim Unif(0.05, 0.45)\)</span>.</li>
<li>Number of SNPs <span class="math inline">\(1000\)</span>.</li>
<li>Heritabillty <span class="math inline">\(h^2 = 0.05\)</span> for all phenotypes.</li>
</ul>
</div>
<div id="simulation-for-association-test" class="section level2">
<h2><span class="header-section-number">2.2</span> Simulation for association test</h2>
<ol style="list-style-type: decimal">
<li>Simulate haplotypes for the father, <span class="math inline">\(H^{f, 1}, H^{f, 2}\)</span> and <span class="math inline">\(H^{m, 1}, H^{m, 2}\)</span>.</li>
<li>Simulate phenotypes for the parents, <span class="math inline">\(y^f\)</span> and <span class="math inline">\(y^m\)</span>.</li>
<li>Let <span class="math inline">\(H^1 = H^{f, 1}\)</span> and <span class="math inline">\(H^2 = H^{m, 1}\)</span>.</li>
</ol>
<p>Some paramter settings:</p>
<pre class="r"><code>step2_sample_size = 10000
step2_nrepeat = 500
step2_maf_low = 0.05
step2_maf_high = 0.45
step2_heritability = 0.001
step2_beta = 1</code></pre>
<ul>
<li>Sample size <span class="math inline">\(10^{4}\)</span>.</li>
<li>Number of replication <span class="math inline">\(500\)</span>.</li>
<li>Monor allele frequency <span class="math inline">\(f \sim Unif(0.05, 0.45)\)</span>.</li>
<li>Heritability (note that it is single SNP scenario) <span class="math inline">\(0.001\)</span>.</li>
</ul>
</div>
<div id="run-simulation" class="section level2">
<h2><span class="header-section-number">2.3</span> Run simulation</h2>
<p>Data for imputation.</p>
<pre class="r"><code>source(&#39;../code/rlib_simulation.R&#39;)</code></pre>
<pre class="r"><code>step1_maf = get_maf(step1_n_snp, step1_maf_low, step1_maf_high)
step1_h_father = sim_hap(step1_sample_size, step1_n_snp, step1_maf)
step1_h_mother = sim_hap(step1_sample_size, step1_n_snp, step1_maf)
step1_h_gwas = sim_hap(step1_sample_size_extra, step1_n_snp, step1_maf)
step1_h_child = transmit_haplo(step1_h_father, step1_h_mother)
step1_effect_size = matrix(spike_and_slab(step1_n_snp * step1_n_pheno, 1 - step1_prior_causal, step1_causal_sigma), nrow = step1_n_snp, ncol = step1_n_pheno)
step1_y_father = simulate_pheno(step1_h_father, step1_effect_size, step1_heritabiltiy, step1_maf)
step1_y_mother = simulate_pheno(step1_h_mother, step1_effect_size, step1_heritabiltiy, step1_maf)
step1_y_gwas = simulate_pheno(step1_h_gwas, step1_effect_size, step1_heritabiltiy, step1_maf)</code></pre>
<p>Data for association test.</p>
<pre class="r"><code>step2_maf = get_maf(step2_nrepeat, step2_maf_low, step2_maf_high)
step2_h_father = sim_hap(step2_sample_size, step2_nrepeat, step2_maf)
step2_h_mother = sim_hap(step2_sample_size, step2_nrepeat, step2_maf)
step2_y_father = simulate_pheno_single_snp(step2_h_father, rep(step2_beta, step2_nrepeat), step2_heritability, step2_maf)
step2_y_mother = simulate_pheno_single_snp(step2_h_mother, rep(step2_beta, step2_nrepeat), step2_heritability, step2_maf)
step2_y_father_null = simulate_pheno_single_snp(step2_h_father, rep(0, step2_nrepeat), step2_heritability, step2_maf, null = TRUE)
step2_y_mother_null = simulate_pheno_single_snp(step2_h_mother, rep(0, step2_nrepeat), step2_heritability, step2_maf, null = TRUE)</code></pre>
</div>
</div>
<div id="haplotype-imputation" class="section level1">
<h1><span class="header-section-number">3</span> Haplotype imputation</h1>
<div id="estimated-prs" class="section level2">
<h2><span class="header-section-number">3.1</span> Estimated PRS</h2>
<pre class="r"><code># build PRS
source(&#39;../code/rlib_gwas.R&#39;)
build_prs = function(geno, pheno, prs_p_cutoff = 0.01) {
  n_snp = ncol(geno)
  n_pheno = ncol(pheno)
  effect_size_prs = matrix(0, nrow = n_snp, ncol = n_pheno)
  prs_p_cutoff = 0.01
  prs_z_cutoff = abs(qnorm(prs_p_cutoff / 2))
  for(pp in 1 : n_pheno) {
    message(&#39;Building phenotype &#39;, pp)
    gwas = run_gwas(geno, pheno[, pp])
    pass_ind = abs(gwas$bhat / gwas$bhat_se) &gt; prs_z_cutoff
    effect_size_prs[, pp][pass_ind] = gwas$bhat[pass_ind]
  }
  return(effect_size_prs)
}
step1_Xgwas = step1_h_gwas[[1]] + step1_h_gwas[[2]]
step1_effect_size_prs = build_prs(step1_Xgwas, step1_y_gwas)</code></pre>
<pre><code>## Building phenotype 1</code></pre>
<pre><code>## Building phenotype 2</code></pre>
<pre><code>## Building phenotype 3</code></pre>
<pre><code>## Building phenotype 4</code></pre>
<pre><code>## Building phenotype 5</code></pre>
<pre><code>## Building phenotype 6</code></pre>
<pre><code>## Building phenotype 7</code></pre>
<pre><code>## Building phenotype 8</code></pre>
<pre><code>## Building phenotype 9</code></pre>
<pre><code>## Building phenotype 10</code></pre>
<pre><code>## Building phenotype 11</code></pre>
<pre><code>## Building phenotype 12</code></pre>
<pre><code>## Building phenotype 13</code></pre>
<pre><code>## Building phenotype 14</code></pre>
<pre><code>## Building phenotype 15</code></pre>
<pre><code>## Building phenotype 16</code></pre>
<pre><code>## Building phenotype 17</code></pre>
<pre><code>## Building phenotype 18</code></pre>
<pre><code>## Building phenotype 19</code></pre>
<pre><code>## Building phenotype 20</code></pre>
<pre><code>## Building phenotype 21</code></pre>
<pre><code>## Building phenotype 22</code></pre>
<pre><code>## Building phenotype 23</code></pre>
<pre><code>## Building phenotype 24</code></pre>
<pre><code>## Building phenotype 25</code></pre>
<pre><code>## Building phenotype 26</code></pre>
<pre><code>## Building phenotype 27</code></pre>
<pre><code>## Building phenotype 28</code></pre>
<pre><code>## Building phenotype 29</code></pre>
<pre><code>## Building phenotype 30</code></pre>
</div>
<div id="run-em" class="section level2">
<h2><span class="header-section-number">3.2</span> Run EM</h2>
<pre class="r"><code>source(&#39;../code/rlib_em.R&#39;)
step1_g1 = step1_h_child[[1]] %*% step1_effect_size_prs
step1_g2 = step1_h_child[[2]] %*% step1_effect_size_prs
collector = list()
for(p in c(2, 5, 10, 15, 20, 25, 30)) {
  o = em_algorithm(step1_y_father[, 1:p], step1_y_mother[, 1:p], step1_g1[, 1:p], step1_g2[, 1:p])
  collector[[length(collector) + 1]] = data.frame(z = o$z_prob_n, num_pheno = p, idx = 1 : length(o$z_prob_n))
}
df_estimated = do.call(rbind, collector)</code></pre>
</div>
</div>
<div id="run-association-test" class="section level1">
<h1><span class="header-section-number">4</span> Run association test</h1>
<pre class="r"><code>source(&#39;../code/rlib_gwas.R&#39;)

dm0 = list(
  X = (step2_h_father[[1]] + step2_h_mother[[1]]) / 2, 
  y = step2_y_father, 
  ynull = step2_y_father_null
)
dm1 = function(z) {
  list(
    X = step2_h_father[[1]] * z + step2_h_mother[[1]] * (1 - z),
    y = step2_y_father,
    ynull = step2_y_father_null
  )
}
dm2 = list(
  X = step2_h_father[[1]],
  y = step2_y_father,
  ynull = step2_y_father_null
)</code></pre>
<p>Actual GWAS run.</p>
<pre class="r"><code>re = list()
ys = c(&#39;y&#39;, &#39;ynull&#39;)
types = c(&#39;beta = 1&#39;, &#39;beta = 0&#39;)

# setting 1
for(i in 1 : length(ys)) {
  o = run_gwas_pairwise(dm0$X, dm0[[ys[i]]])
  re[[length(re) + 1]] = data.frame(b = o$bhat, b_se = o$bhat_se, npheno = 0, type = types[i])
}

# setting 2 
for(npheno in unique(df_estimated$num_pheno)) {
  message(&#39;Working on npheno = &#39;, npheno)
  zvec = df_estimated$z[df_estimated$num_pheno == npheno]
  zmat = matrix(rep(zvec, step2_nrepeat), nrow = step2_sample_size, ncol = step2_nrepeat)
  # setting 2
  dm1_ = dm1(zmat)
  for(i in 1 : length(ys)) {
    o = run_gwas_pairwise(dm1_$X, dm1_[[ys[i]]])
    re[[length(re) + 1]] = data.frame(b = o$bhat, b_se = o$bhat_se, npheno = npheno, type = types[i])
  }
}</code></pre>
<pre><code>## Working on npheno = 2</code></pre>
<pre><code>## Working on npheno = 5</code></pre>
<pre><code>## Working on npheno = 10</code></pre>
<pre><code>## Working on npheno = 15</code></pre>
<pre><code>## Working on npheno = 20</code></pre>
<pre><code>## Working on npheno = 25</code></pre>
<pre><code>## Working on npheno = 30</code></pre>
<pre class="r"><code># setting 3
for(i in 1 : length(ys)) {
  o = run_gwas_pairwise(dm2$X, dm2[[ys[i]]])
  re[[length(re) + 1]] = data.frame(b = o$bhat, b_se = o$bhat_se, npheno = Inf, type = types[i])
}

df_re = do.call(rbind, re) %&gt;% mutate(stat = b / b_se)</code></pre>
</div>
<div id="visualization" class="section level1">
<h1><span class="header-section-number">5</span> Visualization</h1>
<p>Test statistic.</p>
<pre class="r"><code>df_re %&gt;% filter(type == &#39;beta = 1&#39;) %&gt;% 
  ggplot() + geom_boxplot(aes(x = factor(npheno), y = stat)) +
  ggtitle(&#39;Test statistic: Under the alternative&#39;) + th2 +
  theme(legend.position = &#39;bottom&#39;)</code></pre>
<p><img src="impute_and_gwas_files/figure-html/stat-1.png" width="672" /></p>
<pre class="r"><code>df_re %&gt;% filter(type == &#39;beta = 0&#39;) %&gt;% 
  ggplot() + geom_boxplot(aes(x = factor(npheno), y = stat)) +
  ggtitle(&#39;Test statistic: Under the null&#39;) + th2 +
  theme(legend.position = &#39;bottom&#39;)</code></pre>
<p><img src="impute_and_gwas_files/figure-html/stat-2.png" width="672" /></p>
<p>QQ-plot.</p>
<pre class="r"><code>df_re = df_re %&gt;% mutate(pval = zval2pval(stat))  # , df = sample_size - 2
df_re = df_re %&gt;% group_by(npheno, type) %&gt;% mutate(pval_exp = rank(pval) / (n() + 1)) %&gt;% ungroup()
df_re %&gt;% filter(type == &#39;beta = 0&#39;) %&gt;% 
  ggplot() + geom_point(aes(x = -log(pval_exp), y = -log(pval), color = factor(npheno))) +
  ggtitle(&#39;Test statistic: Under the null&#39;) + th2 +
  geom_abline(slope = 1, intercept = 0) + 
  theme(legend.position = &#39;bottom&#39;)</code></pre>
<p><img src="impute_and_gwas_files/figure-html/qq-1.png" width="480" /></p>
<pre class="r"><code>df_re = df_re %&gt;% mutate(pval = zval2pval(stat))  # , df = sample_size - 2
df_re = df_re %&gt;% group_by(npheno, type) %&gt;% mutate(pval_exp = rank(pval) / (n() + 1)) %&gt;% ungroup()
df_re %&gt;% filter(type == &#39;beta = 1&#39;) %&gt;% 
  ggplot() + geom_point(aes(x = -log(pval_exp), y = -log(pval), color = factor(npheno))) +
  ggtitle(&#39;Test statistic: Under the alternative&#39;) + th2 +
  geom_abline(slope = 1, intercept = 0) + 
  theme(legend.position = &#39;bottom&#39;)</code></pre>
<p><img src="impute_and_gwas_files/figure-html/qq2-1.png" width="480" /></p>
<p>Beta hat.</p>
<pre class="r"><code>df_re %&gt;% filter(type == &#39;beta = 1&#39;) %&gt;% 
  ggplot() + geom_boxplot(aes(x = factor(npheno), y = b)) +
  ggtitle(&#39;Beta hat: Under the alternative&#39;) + th2 +
  theme(legend.position = &#39;bottom&#39;) +
  geom_hline(yintercept = 1)</code></pre>
<p><img src="impute_and_gwas_files/figure-html/bhat-1.png" width="672" /></p>
<pre class="r"><code>df_re %&gt;% filter(type == &#39;beta = 0&#39;) %&gt;% 
  ggplot() + geom_boxplot(aes(x = factor(npheno), y = b)) +
  ggtitle(&#39;Beta hat: Under the null&#39;) + th2 +
  theme(legend.position = &#39;bottom&#39;) +
  geom_hline(yintercept = 0)</code></pre>
<p><img src="impute_and_gwas_files/figure-html/bhat-2.png" width="672" /></p>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
