<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>For paper: simulation for PRS-based approach</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Imputing parent of origin of haplotype</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/liangyy/haplotype-po">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">For paper: simulation for PRS-based approach</h1>
<h4 class="date">Last update: Jun 17, 2020</h4>

</div>


<p><span class="math display">\[
\newcommand{\E}{\text{E}}
\newcommand{\diag}{\text{diag}}
\]</span></p>
<pre class="r"><code>rm(list = ls())
library(ggplot2)
library(dplyr)
theme_set(theme_bw(base_size=15))
set.seed(2020)

# load some gists
source(&#39;https://gist.githubusercontent.com/liangyy/43912b3ecab5d10c89f9d4b2669871c9/raw/3ca651cfa53ffccb8422f432561138a46e93710f/my_ggplot_theme.R&#39;)</code></pre>
<div id="simulation-procedure" class="section level1">
<h1><span class="header-section-number">1</span> Simulation procedure</h1>
<ol style="list-style-type: decimal">
<li>Simulate parental genetically determined gene expression <span class="math inline">\(G^{\pi, 1}\)</span> and <span class="math inline">\(G^{\pi, 2}\)</span>.</li>
<li>Simulate the parameter <span class="math inline">\(b\)</span> from
<ul>
<li><span class="math inline">\(b = -0.1\)</span></li>
<li><span class="math inline">\(b = 0.1\)</span></li>
<li><span class="math inline">\(b \sim N(0, 1)\)</span></li>
<li><span class="math inline">\(b \sim \max(0, N(0, 1))\)</span></li>
</ul></li>
<li>Simulate some covariates <span class="math inline">\(C_m \sim N(0, 1)\)</span> with effect size <span class="math inline">\(a_m \sim N(0, 1)\)</span>.</li>
<li>Calculate <span class="math inline">\(y^\pi = (G^{\pi, 1} + G^{\pi, 2}) b + \sum_m a_m C_m + \epsilon^\pi\)</span> where the variance of <span class="math inline">\(\epsilon^\pi\)</span> is set such that <span class="math inline">\(h^2 = \frac{\sigma^2_g}{\sigma^2_g + \sigma^2_e} = 10^{-4}, 0.001, 0.01, 0.05\)</span>.</li>
<li>For the child, haplotype 1 is <span class="math inline">\(G^{f, 1}\)</span> and haplotype 2 is <span class="math inline">\(G^{m, 1}\)</span>.</li>
</ol>
<p>Sample size is set to 300 and the number of phenotypes is set to 500 and the number of covariates is set to 4.</p>
</div>
<div id="simulation" class="section level1">
<h1><span class="header-section-number">2</span> Simulation</h1>
<pre class="r"><code>npheno = 500
nindiv = 300
ncovar = 4
beta_constants = c(-0.1, 0.1)
h2s = c(0.0001, 0.001, 0.01, 0.05)</code></pre>
<pre class="r"><code>get_error_var = function(var_g, h2) {
  var_g / h2 * (1 - h2)
} 
sim_error = function(yy, h2) {
  rnorm(length(yy), sd = sqrt(get_error_var(var(yy), h2)))
}

covar_mat = matrix(rnorm(nindiv * ncovar), nrow = nindiv, ncol = ncovar)
Gf1_mat = matrix(rnorm(nindiv * npheno), nrow = nindiv, ncol = npheno)
Gf2_mat = matrix(rnorm(nindiv * npheno), nrow = nindiv, ncol = npheno)
Gm1_mat = matrix(rnorm(nindiv * npheno), nrow = nindiv, ncol = npheno)
Gm2_mat = matrix(rnorm(nindiv * npheno), nrow = nindiv, ncol = npheno)

beta0 = matrix(rnorm(npheno * ncovar), nrow = ncovar, ncol = npheno)
ycovar = covar_mat %*% beta0


data_list = list()
for(beta in beta_constants) {
  data_list[[as.character(beta)]] = list()
  for(h2 in h2s) {
    yf = (Gf1_mat + Gf2_mat) * beta 
    ym = (Gm1_mat + Gm2_mat) * beta 
    error_f = apply(yf, 2, sim_error, h2 = h2)
    error_m = apply(ym, 2, sim_error, h2 = h2)
    yf = yf + error_f + ycovar
    ym = ym + error_m + ycovar
    data_list[[as.character(beta)]][[as.character(h2)]] = list(yf = yf, ym = ym)
  }
}

beta_rand = rnorm(npheno)
data_list[[&#39;random&#39;]] = list()
for(h2 in h2s) {
  yf = .mat_vec_by_row(Gf1_mat + Gf2_mat, beta_rand, &#39;*&#39;) 
  ym = .mat_vec_by_row(Gm1_mat + Gm2_mat, beta_rand, &#39;*&#39;) 
  error_f = apply(yf, 2, sim_error, h2 = h2)
  error_m = apply(ym, 2, sim_error, h2 = h2)
  yf = yf + error_f + ycovar
  ym = ym + error_m + ycovar
  data_list[[&#39;random&#39;]][[as.character(h2)]] = list(yf = yf, ym = ym)
}

beta_truncated = rnorm(npheno)
beta_truncated[beta_truncated &lt; 0] = 0
data_list[[&#39;truncated&#39;]] = list()
for(h2 in h2s) {
  yf = .mat_vec_by_row(Gf1_mat + Gf2_mat, beta_truncated, &#39;*&#39;) 
  ym = .mat_vec_by_row(Gm1_mat + Gm2_mat, beta_truncated, &#39;*&#39;) 
  error_f = apply(yf, 2, sim_error, h2 = h2)
  error_m = apply(ym, 2, sim_error, h2 = h2)
  yf = yf + error_f + ycovar
  ym = ym + error_m + ycovar
  data_list[[&#39;truncated&#39;]][[as.character(h2)]] = list(yf = yf, ym = ym)
}</code></pre>
</div>
<div id="run-prs-based-imputation" class="section level1">
<h1><span class="header-section-number">3</span> Run PRS-based imputation</h1>
<pre class="r"><code>source(&#39;../code/rlib_em_a_variation.R&#39;)

probz_list = list()
for(beta in names(data_list)) {
  message(&#39;Working on beta = &#39;, beta)
  for(h2 in h2s) {
    for(ng in c(TRUE, FALSE)) {
      o = em_algorithm_a_variation(
        data_list[[as.character(beta)]][[as.character(h2)]]$yf,
        data_list[[as.character(beta)]][[as.character(h2)]]$ym, 
        Gf1_mat, Gm1_mat, covar = covar_mat, maxiter = 100, 
        non_negative = ng
      )
      probz_list[[length(probz_list) + 1]] = data.frame(prob_z = o$z_prob_n, h2 = h2, beta = beta, non_negative = ng)
    }
  }
}</code></pre>
<pre><code>## Working on beta = -0.1</code></pre>
<pre><code>## Working on beta = 0.1</code></pre>
<pre><code>## Working on beta = random</code></pre>
<pre><code>## Working on beta = truncated</code></pre>
<pre class="r"><code>df_z = do.call(rbind, probz_list)</code></pre>
</div>
<div id="visualization" class="section level1">
<h1><span class="header-section-number">4</span> Visualization</h1>
<pre class="r"><code>bin_prob_z = function(x, cuts = c(0.1, 0.4, 0.6, 0.9)) {
  left_most = cuts[1]
  left_mid = cuts[2]
  right_mid = cuts[3]
  right_most = cuts[4]
  o = rep(&#39;&#39;, length(x))
  o[x &lt; left_most] = &#39;strongly wrong&#39;
  o[x &gt;= left_most &amp; x &lt; left_mid] = &#39;mildly wrong&#39;
  o[x &gt;= left_mid &amp; x &lt;= right_mid] = &#39;not sure&#39;
  o[x &gt; right_mid &amp; x &lt;= right_most] = &#39;mildly correct&#39;
  o[x &gt; right_most] = &#39;strongly correct&#39;
  o = factor(o, levels = c(&#39;strongly wrong&#39;, &#39;mildly wrong&#39;, &#39;not sure&#39;, &#39;mildly correct&#39;, &#39;strongly correct&#39;))
  o
}
p = df_z %&gt;% mutate(gamma_category = bin_prob_z(prob_z)) %&gt;% ggplot() + geom_bar(aes(x = gamma_category, fill = non_negative), position = &#39;dodge&#39;) + facet_grid(h2~beta) + theme(legend.position = &#39;bottom&#39;) + theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) + th2 + 
  theme(axis.title.y = element_blank()) + 
  xlab(expression(paste(&#39;Category of &#39;, gamma))) +
  labs(fill = &#39;If b is non-negative&#39;)
ggsave(&#39;../analysis_output/paper_simulation_prs.png&#39;, p, width = 6, height = 6)</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
