<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>For paper: simulation for all settings</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Imputing parent of origin of haplotype</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/liangyy/haplotype-po">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">For paper: simulation for all settings</h1>
<h4 class="date">Last update: Jun 17, 2020</h4>

</div>


<p><span class="math display">\[
\newcommand{\E}{\text{E}}
\]</span></p>
<pre class="r"><code>rm(list = ls())
library(ggplot2)
library(dplyr)
theme_set(theme_bw(base_size=15))
set.seed(2020)

# load some gists
source(&#39;https://gist.githubusercontent.com/liangyy/43912b3ecab5d10c89f9d4b2669871c9/raw/3ca651cfa53ffccb8422f432561138a46e93710f/my_ggplot_theme.R&#39;)</code></pre>
<div id="simulation-procedure" class="section level1">
<h1><span class="header-section-number">1</span> Simulation procedure</h1>
<ol style="list-style-type: decimal">
<li>Simulate parental half-genotypes, <span class="math inline">\(H^{\pi, 1}\)</span> and <span class="math inline">\(H^{\pi, 2}\)</span> for each individual’s parentts. Variants are binary variables sampled independently. Minor allele frequency is <span class="math inline">\(U[0.05, 0.25]\)</span></li>
<li>Simulate effecet size <span class="math inline">\(\beta \sim (1 - \pi_0) \delta_0 + \pi_0 N(0, 1)\)</span> with <span class="math inline">\(\pi_0 = 0.5\)</span></li>
<li>Calculate parental genetic effect <span class="math inline">\(G^{\pi, j} = H^{\pi, j} \beta\)</span></li>
<li>Simulate environmental effect such that heritability <span class="math inline">\(h^2\)</span> is as desired value</li>
<li>Calculate observed parental phenotype <span class="math inline">\(y^\pi = G^{\pi, 1} + G^{\pi, 2} + \epsilon^\pi\)</span></li>
<li>Transmit <span class="math inline">\(H^{f, 1}\)</span> and <span class="math inline">\(H^{m, 1}\)</span> to child <span class="math inline">\(H^1\)</span> and <span class="math inline">\(H^2\)</span></li>
</ol>
</div>
<div id="simulation-code" class="section level1">
<h1><span class="header-section-number">2</span> Simulation code</h1>
<pre class="r"><code># parameters
heritabilities = c(0.01, 0.05, 0.1, 0.25, 0.5)  # c(1e-4, 1e-3, 0.01, 0.05, 0.1, 0.2, 0.5)
maf_range = c(0.05, 0.45)
prior_non_zero = 0.5
nsample = 1000
npheno = 100
nsnp = 50
ngwas = 20000

# functions
.sim_one_snp = function(maf, nindiv) {
  as.numeric(runif(nindiv) &lt; maf)
}
simulate_half_genotype = function(maf, nsnp, nindiv) {
  h1 = sapply(maf, .sim_one_snp, nindiv = nindiv)
  h2 = sapply(maf, .sim_one_snp, nindiv = nindiv)
  return(list(h1 = h1, h2 = h2))
}
y_from_g = function(g_mat, h2) {
  var_g = apply(g_mat, 2, var)
  var_e = (1 - h2) / h2 * var_g
  error = sapply(var_e, function(x) {
    rnorm(nrow(g_mat), sd = sqrt(x))
  })
  return(list(y = g_mat + error, var_e = var_e))
}

# simulation for imputation
maf = runif(nsnp, maf_range[1], maf_range[2])
h_father = simulate_half_genotype(maf, nsnp, nsample)
h_mother = simulate_half_genotype(maf, nsnp, nsample)
beta = matrix(rnorm(npheno * nsnp), ncol = npheno, nrow = nsnp)
g1_father = h_father$h1 %*% beta
g2_father = h_father$h2 %*% beta
g1_mother = h_mother$h1 %*% beta
g2_mother = h_mother$h2 %*% beta
y_father = list()
y_mother = list()
var_e_father = list()
var_e_mother = list()
for(h2 in heritabilities) {
  tmp_f = y_from_g(g1_father + g2_father, h2)
  tmp_m = y_from_g(g1_mother + g2_mother, h2)
  y_father[[as.character(h2)]] = tmp_f$y
  y_mother[[as.character(h2)]] = tmp_m$y
  var_e_father[[as.character(h2)]] = tmp_f$var_e
  var_e_mother[[as.character(h2)]] = tmp_m$var_e
}
hh1 = h_father$h1
hh2 = h_mother$h1
g1 = g1_father
g2 = g1_mother

# simulation for PRS training
h_gwas = simulate_half_genotype(maf, nsnp, ngwas)
x_gwas = h_gwas$h1 + h_gwas$h2
g_gwas = x_gwas %*% beta
y_gwas = list()
for(h2 in heritabilities) {
  y_gwas[[as.character(h2)]] = y_from_g(g_gwas, h2)$y
}</code></pre>
</div>
<div id="prs-fitting" class="section level1">
<h1><span class="header-section-number">3</span> PRS fitting</h1>
<pre class="r"><code>source(&#39;../code/rlib_gwas.R&#39;)

# function
build_prs = function(geno, pheno, prs_p_cutoff = 0.05) {
  n_snp = ncol(geno)
  n_pheno = ncol(pheno)
  effect_size_prs = matrix(0, nrow = n_snp, ncol = n_pheno)
  prs_p_cutoff = 0.01
  prs_z_cutoff = abs(qnorm(prs_p_cutoff / 2))
  for(pp in 1 : n_pheno) {
    # message(&#39;Building phenotype &#39;, pp)
    gwas = run_gwas(geno, pheno[, pp])
    pass_ind = abs(gwas$bhat / gwas$bhat_se) &gt; prs_z_cutoff
    effect_size_prs[, pp][pass_ind] = gwas$bhat[pass_ind]
  }
  return(effect_size_prs)
}

# code
ngwas_to_use = c(0.25, 0.5, 1) * ngwas
prs_model = list()
for(h2 in names(y_gwas)) {
  message(&#39;Working on h2 = &#39;, h2)
  yy = y_gwas[[h2]]
  prs_model[[as.character(h2)]] = list()
  for(nn in ngwas_to_use) {
    message(&#39;Working on nn = &#39;, nn)
    effect_size_prs = build_prs(x_gwas[1:nn, ], yy[1:nn, ])
    prs_model[[as.character(h2)]][[as.character(nn)]] = effect_size_prs
  }
}</code></pre>
<pre><code>## Working on h2 = 0.01</code></pre>
<pre><code>## Working on nn = 5000</code></pre>
<pre><code>## Working on nn = 10000</code></pre>
<pre><code>## Working on nn = 20000</code></pre>
<pre><code>## Working on h2 = 0.05</code></pre>
<pre><code>## Working on nn = 5000</code></pre>
<pre><code>## Working on nn = 10000</code></pre>
<pre><code>## Working on nn = 20000</code></pre>
<pre><code>## Working on h2 = 0.1</code></pre>
<pre><code>## Working on nn = 5000</code></pre>
<pre><code>## Working on nn = 10000</code></pre>
<pre><code>## Working on nn = 20000</code></pre>
<pre><code>## Working on h2 = 0.25</code></pre>
<pre><code>## Working on nn = 5000</code></pre>
<pre><code>## Working on nn = 10000</code></pre>
<pre><code>## Working on nn = 20000</code></pre>
<pre><code>## Working on h2 = 0.5</code></pre>
<pre><code>## Working on nn = 5000</code></pre>
<pre><code>## Working on nn = 10000</code></pre>
<pre><code>## Working on nn = 20000</code></pre>
</div>
<div id="imputation" class="section level1">
<h1><span class="header-section-number">4</span> Imputation</h1>
<pre class="r"><code># number of phenotypes to test
npheno_to_use = c(2, 5, 10, seq(20, 50, by = 10), seq(60, 100, by = 20))

# ideal
slice_m = function(mat, p) {
  mat[, 1 : p, drop = FALSE]
}

sq_diff = function(y, g, s2) {
  rowSums((g - y) ^ 2 / s2 / 2)
}
eval_gamma = function(yf, ym, g1, g2, sigma2f, sigma2m) {
  L1 = sq_diff(yf, g1, sigma2f) + sq_diff(ym, g2, sigma2m)
  L0 = sq_diff(yf, g2, sigma2f) + sq_diff(ym, g1, sigma2m)
  1 / (1 + exp(- (L0 - L1)))
}
df_ideal = list()
for(h2 in names(y_father)) {
  message(&#39;ideal: h2 = &#39;, h2)
  Yf = y_father[[h2]]
  Ym = y_mother[[h2]]
  var_e_f = var_e_father[[h2]]
  var_e_m = var_e_mother[[h2]]
  G1 = g1
  G2 = g2
  # loop over the number of phenotypes
  for(pp in npheno_to_use) {
    # message(pp)
    df_ideal[[length(df_ideal) + 1]] = data.frame(
      gamma = eval_gamma(slice_m(Yf, pp), slice_m(Ym, pp), slice_m(G1, pp), slice_m(G2, pp), var_e_f[1:pp], var_e_m[1:pp]),
      npheno = pp,
      heritability = h2,
      ngwas = NA
    )
  }
}</code></pre>
<pre><code>## ideal: h2 = 0.01</code></pre>
<pre><code>## ideal: h2 = 0.05</code></pre>
<pre><code>## ideal: h2 = 0.1</code></pre>
<pre><code>## ideal: h2 = 0.25</code></pre>
<pre><code>## ideal: h2 = 0.5</code></pre>
<pre class="r"><code>df_ideal = do.call(rbind, df_ideal)

# on-the-fly
source(&#39;../code/rlib_em_otf.R&#39;)
df_otf = list()
for(h2 in names(y_father)) {
  message(&#39;on-the-fly: h2 = &#39;, h2)
  Yf = y_father[[h2]]
  Ym = y_mother[[h2]]
  H1 = hh1
  H2 = hh2
  # loop over the number of phenotypes
  for(pp in npheno_to_use) {
    # message(pp)
    df_otf[[length(df_otf) + 1]] = data.frame(
      gamma = em_algorithm_otf(slice_m(Yf, pp), slice_m(Ym, pp), H1, H2, maxiter = 50)$gamma,
      npheno = pp,
      heritability = h2,
      ngwas = NA
    )
  }
}</code></pre>
<pre><code>## on-the-fly: h2 = 0.01</code></pre>
<pre><code>## on-the-fly: h2 = 0.05</code></pre>
<pre><code>## on-the-fly: h2 = 0.1</code></pre>
<pre><code>## on-the-fly: h2 = 0.25</code></pre>
<pre><code>## on-the-fly: h2 = 0.5</code></pre>
<pre class="r"><code>df_otf = do.call(rbind, df_otf)

# prs-based
source(&#39;../code/rlib_em_a_variation.R&#39;)
df_prs = list()
for(h2 in names(y_father)) {
  message(&#39;prs-based: h2 = &#39;, h2)
  Yf = y_father[[h2]]
  Ym = y_mother[[h2]]
  H1 = hh1
  H2 = hh2
  # loop over prs models
  for(nn in names(prs_model[[h2]])) {
    beta_now = prs_model[[h2]][[nn]]
    prs1 = H1 %*% beta_now
    prs2 = H2 %*% beta_now
    # loop over the number of phenotypes
    for(pp in npheno_to_use) {
      # message(pp)
      df_prs[[length(df_prs) + 1]] = data.frame(
        gamma = em_algorithm_a_variation(slice_m(Yf, pp), slice_m(Ym, pp), slice_m(prs1, pp), slice_m(prs2, pp), maxiter = 50)$z_prob_n,
        npheno = pp,
        heritability = h2,
      ngwas = nn
      )
    }
  }
}</code></pre>
<pre><code>## prs-based: h2 = 0.01</code></pre>
<pre><code>## prs-based: h2 = 0.05</code></pre>
<pre><code>## prs-based: h2 = 0.1</code></pre>
<pre><code>## prs-based: h2 = 0.25</code></pre>
<pre><code>## prs-based: h2 = 0.5</code></pre>
<pre class="r"><code>df_prs = do.call(rbind, df_prs)</code></pre>
</div>
<div id="visualization" class="section level1">
<h1><span class="header-section-number">5</span> Visualization</h1>
<div id="within-prs-based-approach" class="section level2">
<h2><span class="header-section-number">5.1</span> Within PRS-based approach</h2>
<pre class="r"><code>p = df_prs %&gt;% ggplot() + 
  geom_violin(aes(x = factor(npheno), y = gamma, color = ngwas), scale = &#39;width&#39;, position = position_dodge(width = 0.8)) + 
  geom_boxplot(aes(x = factor(npheno), y = gamma, color = ngwas), position = position_dodge(width = 0.8), width = 0.3) + 
  facet_wrap(~heritability, ncol = 2) + 
  th2 + 
  geom_hline(yintercept = 0.5, color = &#39;gray&#39;)
ggsave(&#39;../analysis_output/paper_simulation_all_within_prs.png&#39;, p, height = 7, width = 10)</code></pre>
</div>
<div id="among-all-approaches" class="section level2">
<h2><span class="header-section-number">5.2</span> Among all approaches</h2>
<pre class="r"><code>p = rbind(
  df_ideal %&gt;% mutate(type = &#39;ideal&#39;),
  df_otf %&gt;% mutate(type = &#39;on-the-fly&#39;),
  df_prs %&gt;% filter(ngwas == 10000) %&gt;% mutate(type = &#39;PRS-based&#39;)
) %&gt;% ggplot() + 
  geom_violin(aes(x = factor(npheno), y = gamma, color = type), scale = &#39;width&#39;, position = position_dodge(width = 0.8)) + 
  geom_boxplot(aes(x = factor(npheno), y = gamma, color = type), position = position_dodge(width = 0.8), width = 0.3) + 
  facet_wrap(~heritability, ncol = 2) + 
  th2 +
  geom_hline(yintercept = 0.5, color = &#39;gray&#39;)
ggsave(&#39;../analysis_output/paper_simulation_all_all_approaches.png&#39;, p, height = 7, width = 10)</code></pre>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
