<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>For paper: simulation for the GWAS approahes</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Imputing parent of origin of haplotype</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/liangyy/haplotype-po">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">For paper: simulation for the GWAS approahes</h1>
<h4 class="date">Last update: Jun 19, 2020</h4>

</div>


<p><span class="math display">\[
\newcommand{\E}{\text{E}}
\newcommand{\diag}{\text{diag}}
\]</span></p>
<pre class="r"><code>rm(list = ls())
library(ggplot2)
library(dplyr)
theme_set(theme_bw(base_size=15))
set.seed(2020)

# load some gists
source(&#39;https://gist.githubusercontent.com/liangyy/43912b3ecab5d10c89f9d4b2669871c9/raw/3ca651cfa53ffccb8422f432561138a46e93710f/my_ggplot_theme.R&#39;)</code></pre>
<div id="simulation-procedure" class="section level1">
<h1><span class="header-section-number">1</span> Simulation procedure</h1>
<ol style="list-style-type: decimal">
<li>Simulate haplotype <span class="math inline">\(H^1\)</span> and <span class="math inline">\(H^2\)</span> from Bernoulli where MAF is sampled from <span class="math inline">\(U[0.05, 0.45]\)</span>.</li>
<li>Simulate phenotype <span class="math inline">\(y = H^1 \beta + \epsilon\)</span> with <span class="math inline">\(\beta\)</span> is set to 0 or 1 and per-SNP heritability 0.001 when <span class="math inline">\(\beta = 1\)</span>.</li>
<li>Simulate <span class="math inline">\(\gamma\)</span> from a variaty of distribution:
<ul>
<li>Constant: <span class="math inline">\(\gamma = 0.9\)</span>, <span class="math inline">\(\gamma = 0.1\)</span>, and <span class="math inline">\(\gamma = 0.5\)</span></li>
<li>Beta distribution: Beta(5, 2), Beta(2, 5), and Beta(2, 2)</li>
</ul></li>
</ol>
<p>Sample size 5,000 and simulate 500 phenotype-genotype pairs for each setting.</p>
</div>
<div id="run-simulation" class="section level1">
<h1><span class="header-section-number">2</span> Run simulation</h1>
<pre class="r"><code>sample_size = 5000
nrepeat = 500
maf_range = c(0.05, 0.45)
per_snp_h2 = 0.001
gamma_types = c(&#39;high&#39;, &#39;low&#39;, &#39;middle&#39;)
gamma_constants = data.frame(const = c(0.9, 0.1, 0.5), name = gamma_types, stringsAsFactors = F)
gamma_betas = data.frame(a = c(5, 2, 2), b = c(2, 5, 2), name = gamma_types, stringsAsFactors = F)

# functions
.sim_one_snp = function(maf, nindiv) {
  as.numeric(runif(nindiv) &lt; maf)
}
simulate_half_genotype = function(maf, nsnp, nindiv) {
  sapply(maf, .sim_one_snp, nindiv = nindiv)
}
y_from_g = function(g_mat, h2) {
  var_g = apply(g_mat, 2, var)
  var_e = (1 - h2) / h2 * var_g
  error = sapply(var_e, function(x) {
    rnorm(nrow(g_mat), sd = sqrt(x))
  })
  g_mat + error
}

maf = runif(nrepeat, maf_range[1], maf_range[2])
h1 = simulate_half_genotype(maf, nsnp, sample_size)
h2 = simulate_half_genotype(maf, nsnp, sample_size)
y_alt = y_from_g(h1, per_snp_h2)
y_null = matrix(rnorm(sample_size * nrepeat), ncol = nrepeat, nrow = sample_size)

gamma_list = list()
gamma_list$constant = list()
for(cc in 1 : nrow(gamma_constants)) {
  gamma_list$constant[[gamma_constants$name[cc]]] = matrix(gamma_constants$const[cc], ncol = nrepeat, nrow = sample_size)
}
gamma_list$constant[[&#39;optimal&#39;]] = matrix(1, ncol = nrepeat, nrow = sample_size)

gamma_list$beta = list()
for(cc in 1 : nrow(gamma_betas)) {
  gamma_list$beta[[gamma_betas$name[cc]]] = matrix(rep(rbeta(sample_size, gamma_betas$a[cc], gamma_betas$b[cc]), nrepeat), ncol = nrepeat, nrow = sample_size, byrow = FALSE)
}</code></pre>
</div>
<div id="run-gwas" class="section level1">
<h1><span class="header-section-number">3</span> Run GWAS</h1>
<p>Before running GWAS, we first have some code set up to compute <span class="math inline">\(\bar{r}\)</span> and <span class="math inline">\(\bar{S}\)</span>.</p>
<pre class="r"><code>calc_r = function(zmat) {
  apply(zmat, 2, mean)
}
calc_s = function(zmat) {
  apply(zmat, 2, function(x) {
    mean(x ^ 2 + (1 - x) ^ 2)
  })
}</code></pre>
<pre class="r"><code>source(&#39;../code/rlib_soft_gwas.R&#39;)
source(&#39;../code/rlib_imputed_gwas.R&#39;)

df_gwas = list()
for(gamma_type in names(gamma_list)) {
  for(gg in names(gamma_list[[gamma_type]])) {
    message(&#39;On gamma distr = &#39;, gamma_type, &#39; and gamma type = &#39;, gg)
    gamma_ = gamma_list[[gamma_type]][[gg]]
    mean_r = calc_r(gamma_)
    mean_s = calc_s(gamma_)
    # run soft-gwas 
    ## on alternative
    o_soft = soft_gwas(y_alt, h1, h2, gamma_)
    df_gwas[[length(df_gwas) + 1]] = data.frame(y_type = &#39;alt&#39;, gamma_dist = gamma_type, gamma_type = gg, method = &#39;soft-GWAS&#39;, bhat = o_soft$bhat, se = o_soft$bhat_se, mean_r = mean_r, mean_s = mean_s, idx = 1 : ncol(y_alt), stringsAsFactors = F)
    ## on null
    o_soft = soft_gwas(y_null, h1, h2, gamma_)
    df_gwas[[length(df_gwas) + 1]] = data.frame(y_type = &#39;null&#39;, gamma_dist = gamma_type, gamma_type = gg, method = &#39;soft-GWAS&#39;, bhat = o_soft$bhat, se = o_soft$bhat_se, mean_r = mean_r, mean_s = mean_s, idx = 1 : ncol(y_null), stringsAsFactors = F)
    # run imputed-gwas on alternative
    ## on alternative
    o_imputed = imputed_gwas(y_alt, h1, h2, gamma_)
    df_gwas[[length(df_gwas) + 1]] = data.frame(y_type = &#39;alt&#39;, gamma_dist = gamma_type, gamma_type = gg, method = &#39;imputed-GWAS&#39;, bhat = o_imputed$bhat, se = o_imputed$bhat_se, mean_r = mean_r, mean_s = mean_s, idx = 1 : ncol(y_alt), stringsAsFactors = F)
    ## on null
    o_imputed = imputed_gwas(y_null, h1, h2, gamma_)
    df_gwas[[length(df_gwas) + 1]] = data.frame(y_type = &#39;null&#39;, gamma_dist = gamma_type, gamma_type = gg, method = &#39;imputed-GWAS&#39;, bhat = o_imputed$bhat, se = o_imputed$bhat_se, mean_r = mean_r, mean_s = mean_s, idx = 1 : ncol(y_null), stringsAsFactors = F)
  }
}</code></pre>
<pre><code>## On gamma distr = constant and gamma type = high</code></pre>
<pre><code>## On gamma distr = constant and gamma type = low</code></pre>
<pre><code>## On gamma distr = constant and gamma type = middle</code></pre>
<pre><code>## On gamma distr = constant and gamma type = optimal</code></pre>
<pre><code>## On gamma distr = beta and gamma type = high</code></pre>
<pre><code>## On gamma distr = beta and gamma type = low</code></pre>
<pre><code>## On gamma distr = beta and gamma type = middle</code></pre>
<pre class="r"><code>df_gwas = do.call(rbind, df_gwas)
df_gwas = df_gwas %&gt;% mutate(pval = pnorm(abs(bhat / se), lower.tail = F) * 2, test_stat = bhat / se)
df_gwas = df_gwas %&gt;% group_by(y_type, gamma_dist, gamma_type, method) %&gt;% mutate(p_expected = rank(pval) / (1 + n())) %&gt;% ungroup()
df_gwas$method = factor(df_gwas$method, levels = c(&#39;soft-GWAS&#39;, &#39;imputed-GWAS&#39;))
df_gwas$gamma_type = factor(df_gwas$gamma_type, levels = c(&#39;low&#39;, &#39;middle&#39;, &#39;high&#39;, &#39;optimal&#39;))
df_gwas$gamma_dist = factor(df_gwas$gamma_dist, levels = c(&#39;constant&#39;, &#39;beta&#39;))
df_gwas$y_type = factor(df_gwas$y_type, levels = c(&#39;null&#39;, &#39;alt&#39;))
df_stat = df_gwas %&gt;%  filter(y_type == &#39;alt&#39;) %&gt;% group_by(y_type, gamma_dist, gamma_type, method) %&gt;% summarize(power = unique(mean_r) / unique(sqrt(mean_s)), bias = unique(mean_r) / unique(mean_s))</code></pre>
</div>
<div id="visualization" class="section level1">
<h1><span class="header-section-number">4</span> Visualization</h1>
<div id="qq-plot-under-the-null" class="section level2">
<h2><span class="header-section-number">4.1</span> QQ plot under the null</h2>
<pre class="r"><code>p = df_gwas %&gt;% filter(y_type == &#39;null&#39;) %&gt;% ggplot() + geom_point(aes(x = -log(p_expected), y = -log(pval), color = gamma_type)) + facet_grid(gamma_dist~method) + geom_abline(intercept = 0, slope = 1, color = &#39;black&#39;, linetype = 2) + th2 + coord_equal() + theme(legend.title = element_blank(), legend.position = &#39;right&#39;) + xlab(expression(-log(p[expected]))) + ylab(expression(-log(p[observed])))  ;p</code></pre>
<p><img src="paper_simulation_gwas_files/figure-html/vis%20qqplot-1.png" width="480" /></p>
<pre class="r"><code>ggsave(&#39;../analysis_output/paper_simulation_gwas_qqplot.png&#39;, p, width = 5, height = 4)</code></pre>
</div>
<div id="effect-size-under-the-alternative" class="section level2">
<h2><span class="header-section-number">4.2</span> Effect size under the alternative</h2>
<pre class="r"><code>p = df_gwas %&gt;% filter(y_type == &#39;alt&#39;) %&gt;% ggplot() + 
  geom_violin(aes(x = gamma_type, y = bhat), position = position_dodge(width = 0.5)) + 
  geom_boxplot(aes(x = gamma_type, y = bhat), position = position_dodge(width = 0.5), width = 0.3) + 
  facet_grid(gamma_dist~method) + 
  geom_point(data = df_stat, aes(x = gamma_type, y = bias), shape = 4, color = &#39;red&#39;, size = 3) + 
  th2 + 
  xlab(expression(paste(&#39;Type of &#39;, gamma))) + 
  ylab(expression(paste(&#39;Estimated &#39;, beta))); p</code></pre>
<p><img src="paper_simulation_gwas_files/figure-html/vis%20effect%20size-1.png" width="480" /></p>
<pre class="r"><code>ggsave(&#39;../analysis_output/paper_simulation_gwas_bhat.png&#39;, p, width = 5, height = 4)</code></pre>
</div>
<div id="test-statistic" class="section level2">
<h2><span class="header-section-number">4.3</span> Test statistic</h2>
<pre class="r"><code>df_opt = df_gwas %&gt;% filter(y_type == &#39;alt&#39;, gamma_type == &#39;optimal&#39;)
df_power = df_gwas %&gt;% filter(y_type == &#39;alt&#39;, gamma_type != &#39;optimal&#39;) %&gt;% left_join(df_opt %&gt;% select(y_type, method, idx, test_stat), by = c(&#39;y_type&#39;, &#39;method&#39;, &#39;idx&#39;), suffix = c(&#39;&#39;, &#39;.optimal&#39;)) %&gt;% mutate(test_stat_ratio = test_stat / test_stat.optimal) 
p = df_power %&gt;% ggplot() +
   geom_violin(aes(x = gamma_type, y = test_stat_ratio), position = position_dodge(width = 0.5), scale = &#39;width&#39;) + 
  geom_boxplot(aes(x = gamma_type, y = test_stat_ratio), position = position_dodge(width = 0.5), width = 0.3) + 
  facet_grid(gamma_dist~method) + 
  geom_point(data = df_stat %&gt;% filter(gamma_type != &#39;optimal&#39;), aes(x = gamma_type, y = power), shape = 4, color = &#39;red&#39;, size = 3) + 
  th2 +
  ylim(c(-5, 5)) +
  xlab(expression(paste(&#39;Type of &#39;, gamma))) + 
  ylab(expression(paste(&#39;Ratio of test statistic &#39;, T / T^&quot;*&quot;))); p</code></pre>
<pre><code>## Warning: Removed 68 rows containing non-finite values (stat_ydensity).</code></pre>
<pre><code>## Warning: Removed 68 rows containing non-finite values (stat_boxplot).</code></pre>
<p><img src="paper_simulation_gwas_files/figure-html/test-1.png" width="384" /></p>
<pre class="r"><code>ggsave(&#39;../analysis_output/paper_simulation_gwas_test_stat.png&#39;, p, height = 4, width = 4)</code></pre>
<pre><code>## Warning: Removed 68 rows containing non-finite values (stat_ydensity).

## Warning: Removed 68 rows containing non-finite values (stat_boxplot).</code></pre>
<pre class="r"><code>p = df_power %&gt;% group_by(method, gamma_type, gamma_dist) %&gt;% summarize(median_ratio = median(test_stat_ratio), mean_ratio = mean(test_stat_ratio)) %&gt;% left_join(df_stat, by = c(&#39;method&#39;, &#39;gamma_type&#39;, &#39;gamma_dist&#39;)) %&gt;% 
  ggplot() + geom_point(aes(x = power, y = median_ratio, color = &#39;median of ratio&#39;, shape = method), size = 3, alpha = 0.5) + geom_point(aes(x = power, y = mean_ratio, color = &#39;mean of ratio&#39;, shape = method), size = 3, alpha = 0.5) + th + geom_abline(intercept = 0, slope = 1, linetype = 2) + coord_equal() + theme(legend.title = element_blank(), legend.position = c(0.25, 0.8), legend.margin = margin(0)) +
  xlab(expression(paste(&#39;Theoretical &#39;, T / T^&quot;*&quot;))) + 
  ylab(expression(paste(&#39;Observed &#39;, T / T^&quot;*&quot;))); p</code></pre>
<p><img src="paper_simulation_gwas_files/figure-html/test2-1.png" width="384" /></p>
<pre class="r"><code>ggsave(&#39;../analysis_output/paper_simulation_gwas_test_stat_scatter.png&#39;, p, height = 4, width = 4)</code></pre>
</div>
<div id="compare-soft-gwas-and-imputed-gwas" class="section level2">
<h2><span class="header-section-number">4.4</span> Compare soft-GWAS and imputed-GWAS</h2>
<pre class="r"><code>df_gwas_compare = inner_join(
  df_gwas %&gt;% filter(method == &#39;soft-GWAS&#39;, y_type == &#39;alt&#39;),
  df_gwas %&gt;% filter(method == &#39;imputed-GWAS&#39;, y_type == &#39;alt&#39;),
  by = c(&#39;y_type&#39;, &#39;gamma_dist&#39;, &#39;gamma_type&#39;, &#39;idx&#39;),
  suffix = c(&#39;_soft&#39;, &#39;_imputed&#39;)
)
p = df_gwas_compare %&gt;% ggplot() + 
  geom_point(aes(x = bhat_soft, y = bhat_imputed)) + 
  facet_grid(gamma_dist~gamma_type) + th2 + geom_abline(slope = 1, intercept = 0, linetype = 2, color = &#39;gray&#39;) +
  xlab(expression(paste(&#39;Estimated &#39;, beta, &#39; in soft-GWAS&#39;))) +
  ylab(expression(paste(&#39;Estimated &#39;, beta, &#39; in imputed-GWAS&#39;))); p</code></pre>
<p><img src="paper_simulation_gwas_files/figure-html/compare-1.png" width="576" /></p>
<pre class="r"><code>ggsave(&#39;../analysis_output/paper_simulation_gwas_compare_bhat.png&#39;, p, height = 3.5, width = 6)

p = df_gwas_compare %&gt;% ggplot() + 
  geom_point(aes(x = se_soft, y = se_imputed)) + 
  facet_grid(gamma_dist~gamma_type) + th2 + geom_abline(slope = 1, intercept = 0, linetype = 2, color = &#39;gray&#39;) +
  xlab(expression(paste(&#39;S.E. of &#39;, beta, &#39; in soft-GWAS&#39;))) +
  ylab(expression(paste(&#39;S.E. of &#39;, beta, &#39; in imputed-GWAS&#39;))); p</code></pre>
<p><img src="paper_simulation_gwas_files/figure-html/compare-2.png" width="576" /></p>
<pre class="r"><code>ggsave(&#39;../analysis_output/paper_simulation_gwas_compare_se.png&#39;, p, height = 3.5, width = 6)

p = df_gwas_compare %&gt;% ggplot() + 
  geom_point(aes(x = test_stat_soft, y = test_stat_imputed)) + 
  facet_grid(gamma_dist~gamma_type) + th2 + geom_abline(slope = 1, intercept = 0, linetype = 2, color = &#39;gray&#39;) +
  xlab(expression(paste(&#39;Test statistic &#39;, T, &#39; in soft-GWAS&#39;))) +
  ylab(expression(paste(&#39;Test statistic &#39;, T, &#39; in imputed-GWAS&#39;))); p</code></pre>
<p><img src="paper_simulation_gwas_files/figure-html/compare-3.png" width="576" /></p>
<pre class="r"><code>ggsave(&#39;../analysis_output/paper_simulation_gwas_compare_test_stat.png&#39;, p, height = 3.5, width = 6)</code></pre>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
