<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Second idea</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Imputing parent of origin of haplotype</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/liangyy/haplotype-po">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Second idea</h1>
<h4 class="date">Last update: May 07, 2020</h4>

</div>


<p><span class="math display">\[
\newcommand{\E}{\text{E}}
\newcommand{\diag}{\text{diag}}
\]</span></p>
<pre class="r"><code>library(dplyr)
library(ggplot2)</code></pre>
<div id="about" class="section level1">
<h1><span class="header-section-number">1</span> About</h1>
<p>As tested on real data, the imputation scheme proposed <a href="init_idea_outline.html">here</a> do not get good-quality imputation on real data. As the second try, I propose the following imputation scheme which do not rely on pre-compuated PRS/GWAS results. Instead, it fits a PRS on-the-fly.</p>
<p>In the following, I first sketch the model and then spread out the corresponding EM algorithm.</p>
</div>
<div id="model" class="section level1">
<h1><span class="header-section-number">2</span> Model</h1>
<p>Still, the focus is <span class="math inline">\(y^f, y^m | Z\)</span> where <span class="math inline">\(Z\)</span> is indicator for event that haplotype 1 is from father. Previously, we use <span class="math inline">\(g^1\)</span> and <span class="math inline">\(g^2\)</span> to present the different contributions of the two haplotypes. Here we work with <span class="math inline">\(H^1\)</span> and <span class="math inline">\(H^2\)</span> directly. Specifically, the model is as follow.</p>
<p><span class="math display">\[\begin{aligned}
  y^f | H^1, H^2, Z &amp;\sim N(~[ Z H^1 + (1 - Z) H^2 ]~\beta_f~, \sigma_f^2) \\
  y^m | H^1, H^2, Z &amp;\sim N(~[ (1 - Z) H^1 + Z H^2 ]~\beta_m~, \sigma_f^2)
\end{aligned}\]</span></p>
<p>Notice that I used sex-specific model so that <span class="math inline">\(\beta\)</span> and <span class="math inline">\(\sigma^2\)</span> are sex-specific.</p>
</div>
<div id="em-algorithm-derivation-sketch" class="section level1">
<h1><span class="header-section-number">3</span> EM algorithm (derivation sketch)</h1>
<ul>
<li>Complete log likelihood <span class="math inline">\(\log \Pr(y, Z | H; \beta, \sigma^2)\)</span></li>
</ul>
<p><span class="math display">\[\begin{aligned}
  \log \Pr(y^f, y^m, Z | H^1, H^2; \beta, \sigma^2 ) &amp;= constant \\
  &amp;+ \frac{n}{2} \log \sigma_f^2 + \frac{n}{2} \log \sigma_m^2 \\ 
  &amp;- \frac{1}{2 \sigma_f^2} (y^f - H_Z \beta_f)&#39; (y^f - H_Z \beta_f) \\
  &amp;- \frac{1}{2 \sigma_m^2} (y^m - H_{-Z} \beta_m)&#39;(y^m - H_{-Z} \beta_m)
\end{aligned}\]</span> , where <span class="math inline">\(H_Z := Z H^1 + (1 - Z) H^2\)</span> and <span class="math inline">\(H_{-Z} := (1 - Z) H^1 + Z H^2\)</span>.</p>
<ul>
<li><span class="math inline">\(Q(\theta, \theta^{(t)})\)</span>. Some simplification come in.</li>
</ul>
<p><span class="math display">\[\begin{aligned}
  \E_{Z | y, H, \theta^{(t)}}[\log \Pr(y, Z | H; \theta)] &amp;= constant \\
  &amp;~~+ \frac{n}{2} \log \sigma_f^2 + \frac{1}{\sigma_f^2} \E[ (y^f)&#39; H_Z \beta_f ] - \frac{1}{2 \sigma_f^2} \E[ \beta_f&#39; H_Z&#39; H_Z \beta_f] \\
  &amp;~~+ \cdots \text{counterpart in mother} \\
  &amp;= \frac{n}{2} \log \sigma_f^2 + \frac{1}{\sigma_f^2} (y^f)&#39; [\E(Z) H^1 + (1 - \E(Z)) H^2] \beta_f \\
  &amp;~~- \frac{1}{2 \sigma_f^2} \beta_f&#39; \E[ (H^1)&#39; Z^2 H^1 + 2 (H^1)&#39; Z(1-Z) H^2 + (H^2)&#39;(1 - Z)^2H^2 ]\beta_f \\
  &amp;~~+ \cdots \\
  &amp;~~~~\text{, since $Z^2 = Z$, $(1 - Z)^2 = (1 - Z)$, and $Z(1 - Z) = 0$} \\
  &amp;~~~~\text{, and let $\E_{Z|y, H, \theta^{(t)}}(Z) = \gamma$, a diagonal matrix} \\
  &amp;= \frac{n}{2} \log \sigma_f^2 + \frac{1}{\sigma_f^2} (y^f)&#39; [\gamma H^1 + (1 - \gamma) H^2] \beta_f \\
  &amp;~~- \frac{1}{2 \sigma_f^2} \beta_f&#39; [ (H^1)&#39; \gamma H^1 + (H^2)&#39;(1 - \gamma) H^2 ]\beta_f \\
  &amp;~~+ \cdots 
\end{aligned}\]</span></p>
<ul>
<li><strong>M step</strong> (performed phenotype-by-phenotype)</li>
</ul>
<p>For father, let <span class="math inline">\(M = \gamma H^1 + (1 - \gamma) H^2\)</span> and <span class="math inline">\(N = (H^1)&#39; \gamma H^1 + (H^2)&#39;(1 - \gamma) H^2\)</span>.</p>
<p><span class="math display">\[\begin{aligned}
  \beta^f, \sigma_f^2 &amp;= \arg\max_{\beta, \sigma^2} -\frac{n}{2} \log \sigma^2 + \frac{1}{\sigma^2}  (y^f)&#39; M \beta - \frac{1}{2\sigma^2} \beta&#39; N \beta - \frac{(y^f)&#39; y^f}{2\sigma^2}\\
  \beta^f &amp;= N^{-1} M&#39; y^f \\
  \sigma_f^2 &amp;= \frac{1}{n} \diag[~ (y^f)&#39; y^f - (\beta^f)&#39; N \beta^f ~]  ~~ \text{, given $\beta^f$ is achieved optimum}
\end{aligned}\]</span></p>
<p>For the counterpart in mother, let <span class="math inline">\(\widetilde{M} = (1 - \gamma) H^1 + \gamma H^2 = H^1 + H^2 - M\)</span> and <span class="math inline">\(\widetilde{N} = (H^1)&#39; (1 - \gamma) H^1 + (H^2)&#39;\gamma H^2 = (H^1)&#39;H^1 + (H^2)&#39;H^2 - N\)</span>.</p>
<p><span class="math display">\[\begin{aligned}
  \beta^m &amp;= \widetilde{N}^{-1} \widetilde{M}&#39; y^f \\
  \sigma_m^2 &amp;= \frac{1}{n} \diag[~ (y^m)&#39; y^m - (\beta^m)&#39; \widetilde{N} \beta^m ~]  ~~ \text{, given $\beta^m$ is achieved optimum}
\end{aligned}\]</span></p>
<ul>
<li><strong>E step</strong> (calculate <span class="math inline">\(\gamma\)</span> individual-by-individual)</li>
</ul>
<p><span class="math display">\[\begin{aligned}
  \Pr(Z = z | y^f, y^m, H^1, H^2; \theta) &amp;= \frac{\Pr(y^f, y^m| Z = z, H^1, H^2; \theta)\Pr(Z = z)}{\sum_{z&#39; \in \{0, 1\}} \Pr(y^f, y^m| Z = z&#39;, H^1, H^2; \theta)\Pr(Z = z&#39;)} \\
  \Pr(y^f, y^m| Z = z, H^1, H^2; \theta) &amp;= \prod_p [~ \Pr(y^{f, p} | Z = z, H^1, H^2; \theta^{f, p}) \times \Pr(y^{m, p} | Z = z, H^1, H^2; \theta^{m, p}) ~]
\end{aligned}\]</span></p>
<p>We’d better work in log-scale.</p>
</div>
<div id="em-algorithm-implementation-note" class="section level1">
<h1><span class="header-section-number">4</span> EM algorithm (implementation note)</h1>
<ol style="list-style-type: decimal">
<li>Initialize <span class="math inline">\(\theta^{(0)} = \beta, \sigma^2\)</span> for each sex and phenotype.</li>
<li>E step, calculate <span class="math inline">\(\gamma\)</span> by</li>
</ol>
<p><span class="math display">\[\begin{aligned}
  l_i(z) &amp;= \sum_p \sum_{f, m} \log(y | Z = z, H, \theta^{(t)}) \\
  \gamma_i &amp;= \begin{cases}
    \frac{1}{1 + e^{r_i}} &amp; l_i(1) &gt; l_i(0), r_i = l_i(0) - l_i(1)  &lt; 0 \\
    \frac{e^{r_i}}{1 + e^{r_i}} &amp; l_i(1) \le l_i(0), r_i = l_i(1) - l_i(0) &lt; 0
  \end{cases}
\end{aligned}\]</span>.</p>
<ol start="3" style="list-style-type: decimal">
<li>M step.</li>
</ol>
<p>Let</p>
<p><span class="math display">\[\begin{aligned}
  M &amp;= \diag(\gamma_i) H^1 + (I - \diag(\gamma_i)) H^2 \\
  N &amp;= (H^1)&#39; \diag(\gamma_i) H^1 + (H^2)&#39; (I - \diag(\gamma_i)) H^2 \\
  \widetilde{M} &amp;= H^1 + H^2 - M \\
  \widetilde{N} &amp;= (H^1)&#39;H^1 + (H^2)&#39;H^2 - N
\end{aligned}\]</span>.</p>
<p>And update <span class="math inline">\(\theta\)</span> as</p>
<p><span class="math display">\[\begin{aligned}
  \theta^{(t+1)}: &amp; \\
  \beta^f &amp;= N^{-1} M&#39; y^f \\
  \sigma_f^2 &amp;= \frac{1}{n} \diag[~ (y^f)&#39; y^f - (\beta^f)&#39; N \beta^f ~] \\
  \beta^f &amp;= \widetilde{N}^{-1} \widetilde{M}&#39; y^m \\
  \sigma_m^2 &amp;= \frac{1}{n} \diag[~ (y^m)&#39; y^m - (\beta^m)&#39; \widetilde{N} \beta^m ~]
\end{aligned}\]</span>.</p>
<ol start="4" style="list-style-type: decimal">
<li>Repeat step 2 and 3 until converge.</li>
</ol>
</div>
<div id="implemtation" class="section level1">
<h1><span class="header-section-number">5</span> Implemtation</h1>
<pre class="r"><code>source(&#39;../code/rlib_em_otf.R&#39;)</code></pre>
</div>
<div id="simulation" class="section level1">
<h1><span class="header-section-number">6</span> Simulation</h1>
<p>Now I simulate some data to test the EM implementation.</p>
<p>Mostly reuse the scheme in <a href="init_idea_outline.html">init_idea_outline</a>. It contains the following steps:</p>
<ol style="list-style-type: decimal">
<li>Simulate haplotypes for the parents, <span class="math inline">\(H_i^{f, 1}, H_i^{f, 2}\)</span> and <span class="math inline">\(H_i^{m, 1}, H_i^{m, 2}\)</span>.</li>
<li>Simulate haplotypes for the child, <span class="math inline">\(H_i^1, H_i^2\)</span>.</li>
<li>Simulate phenotypes for the parents, <span class="math inline">\(y_{i, p}^f\)</span> and <span class="math inline">\(y_{i, p}^m\)</span>.</li>
</ol>
<p>Some parameter settings:</p>
<pre class="r"><code>sample_size = 10000
n_pheno = 30
prior_causal = 0.4
causal_sigma = 1
n_snp = 10
heritabiltiy = 0.01
maf_low = 0.05
maf_high = 0.45</code></pre>
<ul>
<li>Sample size <span class="math inline">\(10^{4}\)</span>.</li>
<li>Number of phenotypes <span class="math inline">\(30\)</span>.</li>
<li>Effect size <span class="math inline">\(\beta_{p, k} \sim \pi \delta_0 + (1 - \pi) N(0, \sigma)\)</span> where <span class="math inline">\(\pi = 0.6\)</span> and <span class="math inline">\(\sigma^2 = 1\)</span>.</li>
<li>Allele frequency <span class="math inline">\(maf \sim Unif(0.05, 0.45)\)</span>.</li>
<li>Number of SNPs <span class="math inline">\(10\)</span>.</li>
<li>Heritabillty <span class="math inline">\(h^2 = 0.01\)</span> for all phenotypes.</li>
</ul>
<p>Simulation.</p>
<pre class="r"><code>source(&#39;../code/rlib_simulation.R&#39;)</code></pre>
<pre class="r"><code>maf = get_maf(n_snp, maf_low, maf_high)
h_father = sim_hap(sample_size, n_snp, maf)
h_mother = sim_hap(sample_size, n_snp, maf)
h_child = transmit_haplo(h_father, h_mother)
effect_size = matrix(spike_and_slab(n_snp * n_pheno, 1 - prior_causal, causal_sigma), nrow = n_snp, ncol = n_pheno)
y_father = simulate_pheno(h_father, effect_size, heritabiltiy, maf)
y_mother = simulate_pheno(h_mother, effect_size, heritabiltiy, maf)</code></pre>
</div>
<div id="run-em" class="section level1">
<h1><span class="header-section-number">7</span> Run EM</h1>
<pre class="r"><code>h1 = h_child[[1]]
h2 = h_child[[2]]
df_z = list()
for(np in c(2, 5, 10, 15, 20, 25, 30)) {
  message(&#39;Working on num_pheno = &#39;, np)
  o = em_algorithm_otf(y_father[, 1 : np], y_mother[, 1 : np], h1, h2, maxiter = 50)
  df_z[[length(df_z) + 1]] = data.frame(prob_z = o$gamma, num_pheno = np)
}</code></pre>
<pre><code>## Working on num_pheno = 2</code></pre>
<pre><code>## Working on num_pheno = 5</code></pre>
<pre><code>## Working on num_pheno = 10</code></pre>
<pre><code>## Working on num_pheno = 15</code></pre>
<pre><code>## Working on num_pheno = 20</code></pre>
<pre><code>## Working on num_pheno = 25</code></pre>
<pre><code>## Working on num_pheno = 30</code></pre>
<pre class="r"><code>df_z = do.call(rbind, df_z)</code></pre>
<pre class="r"><code>df_z %&gt;% ggplot() + geom_jitter(aes(x = num_pheno, y = prob_z), height = 0, width = 0.5, alpha = 0.05)</code></pre>
<p><img src="second_idea_outline_files/figure-html/vis-1.png" width="672" /></p>
<pre class="r"><code>df_z %&gt;% group_by(num_pheno) %&gt;% summarize(S_bar = mean(prob_z ^ 2 + (1 - prob_z) ^ 2), gamma_bar = mean(prob_z)) %&gt;% mutate(power_ratio = gamma_bar / sqrt(S_bar))</code></pre>
<pre><code>## # A tibble: 7 x 4
##   num_pheno S_bar gamma_bar power_ratio
##       &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;       &lt;dbl&gt;
## 1         2 0.506     0.505       0.710
## 2         5 0.516     0.512       0.712
## 3        10 0.534     0.524       0.717
## 4        15 0.545     0.534       0.724
## 5        20 0.557     0.544       0.729
## 6        25 0.571     0.556       0.736
## 7        30 0.582     0.566       0.742</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
